{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fbb27a01-0d7d-4ffc-ae36-8665b7468353",
   "metadata": {},
   "source": [
    "# Supervised baseline: RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6ccfd3f4-893a-4386-8b95-0f3c822da406",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "87c0dc4c-ae9f-4e2c-8b70-0d1b84e30012",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from sklearn.feature_extraction.text import CountVectorizer\n",
    "from sklearn.feature_extraction.text import TfidfTransformer\n",
    "from sklearn.naive_bayes import MultinomialNB\n",
    "from sklearn.linear_model import SGDClassifier # Linear SVM.\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn import metrics\n",
    "from sklearn.model_selection import GridSearchCV"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01f2383b-33eb-42a1-9eff-339e12ee6bc7",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f5e19087-22a0-497b-acd8-b583add73aa8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1656 entries, 0 to 1655\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              1656 non-null   int64 \n",
      " 1   Title             1656 non-null   object\n",
      " 2   Authors           1655 non-null   object\n",
      " 3   Citation          1656 non-null   object\n",
      " 4   First Author      1655 non-null   object\n",
      " 5   Journal/Book      1656 non-null   object\n",
      " 6   Publication Year  1656 non-null   int64 \n",
      " 7   Create Date       1656 non-null   object\n",
      " 8   PMCID             943 non-null    object\n",
      " 9   NIHMS ID          149 non-null    object\n",
      " 10  DOI               1656 non-null   object\n",
      " 11  Label             1656 non-null   int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 155.4+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21205307</td>\n",
       "      <td>time-series clustering of gene expression in i...</td>\n",
       "      <td>Ghandhi SA, Sinha A, Markatou M, Amundson SA.</td>\n",
       "      <td>BMC Genomics. 2011 Jan 4;12:2. doi: 10.1186/14...</td>\n",
       "      <td>Ghandhi SA</td>\n",
       "      <td>BMC Genomics</td>\n",
       "      <td>2011</td>\n",
       "      <td>2011/01/06</td>\n",
       "      <td>PMC3022823</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/1471-2164-12-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25409906</td>\n",
       "      <td>dynamic analyses of alternative polyadenylatio...</td>\n",
       "      <td>Xia Z, Donehower LA, Cooper TA, Neilson JR, Wh...</td>\n",
       "      <td>Nat Commun. 2014 Nov 20;5:5274. doi: 10.1038/n...</td>\n",
       "      <td>Xia Z</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>2014</td>\n",
       "      <td>2014/11/21</td>\n",
       "      <td>PMC4467577</td>\n",
       "      <td>NIHMS674338</td>\n",
       "      <td>10.1038/ncomms6274</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14984408</td>\n",
       "      <td>temporal and spatial patterns of kv1.1 and kv1...</td>\n",
       "      <td>Karimi-Abdolrezaee S, Eftekharpour E, Fehlings...</td>\n",
       "      <td>Eur J Neurosci. 2004 Feb;19(3):577-89. doi: 10...</td>\n",
       "      <td>Karimi-Abdolrezaee S</td>\n",
       "      <td>Eur J Neurosci</td>\n",
       "      <td>2004</td>\n",
       "      <td>2004/02/27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1111/j.0953-816x.2004.03164.x</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19340919</td>\n",
       "      <td>simultaneously segmenting multiple gene expres...</td>\n",
       "      <td>Tadepalli S, Ramakrishnan N, Watson LT, Mishra...</td>\n",
       "      <td>J Bioinform Comput Biol. 2009 Apr;7(2):339-56....</td>\n",
       "      <td>Tadepalli S</td>\n",
       "      <td>J Bioinform Comput Biol</td>\n",
       "      <td>2009</td>\n",
       "      <td>2009/04/03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1142/s0219720009004114</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18547802</td>\n",
       "      <td>rna dynamics: it is about time</td>\n",
       "      <td>Al-Hashimi HM, Walter NG.</td>\n",
       "      <td>Curr Opin Struct Biol. 2008 Jun;18(3):321-9. d...</td>\n",
       "      <td>Al-Hashimi HM</td>\n",
       "      <td>Curr Opin Struct Biol</td>\n",
       "      <td>2008</td>\n",
       "      <td>2008/06/13</td>\n",
       "      <td>PMC2580758</td>\n",
       "      <td>NIHMS74438</td>\n",
       "      <td>10.1016/j.sbi.2008.04.004</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  21205307  time-series clustering of gene expression in i...   \n",
       "1  25409906  dynamic analyses of alternative polyadenylatio...   \n",
       "2  14984408  temporal and spatial patterns of kv1.1 and kv1...   \n",
       "3  19340919  simultaneously segmenting multiple gene expres...   \n",
       "4  18547802                     rna dynamics: it is about time   \n",
       "\n",
       "                                             Authors  \\\n",
       "0      Ghandhi SA, Sinha A, Markatou M, Amundson SA.   \n",
       "1  Xia Z, Donehower LA, Cooper TA, Neilson JR, Wh...   \n",
       "2  Karimi-Abdolrezaee S, Eftekharpour E, Fehlings...   \n",
       "3  Tadepalli S, Ramakrishnan N, Watson LT, Mishra...   \n",
       "4                          Al-Hashimi HM, Walter NG.   \n",
       "\n",
       "                                            Citation          First Author  \\\n",
       "0  BMC Genomics. 2011 Jan 4;12:2. doi: 10.1186/14...            Ghandhi SA   \n",
       "1  Nat Commun. 2014 Nov 20;5:5274. doi: 10.1038/n...                 Xia Z   \n",
       "2  Eur J Neurosci. 2004 Feb;19(3):577-89. doi: 10...  Karimi-Abdolrezaee S   \n",
       "3  J Bioinform Comput Biol. 2009 Apr;7(2):339-56....           Tadepalli S   \n",
       "4  Curr Opin Struct Biol. 2008 Jun;18(3):321-9. d...         Al-Hashimi HM   \n",
       "\n",
       "              Journal/Book  Publication Year Create Date       PMCID  \\\n",
       "0             BMC Genomics              2011  2011/01/06  PMC3022823   \n",
       "1               Nat Commun              2014  2014/11/21  PMC4467577   \n",
       "2           Eur J Neurosci              2004  2004/02/27         NaN   \n",
       "3  J Bioinform Comput Biol              2009  2009/04/03         NaN   \n",
       "4    Curr Opin Struct Biol              2008  2008/06/13  PMC2580758   \n",
       "\n",
       "      NIHMS ID                               DOI  Label  \n",
       "0          NaN            10.1186/1471-2164-12-2      1  \n",
       "1  NIHMS674338                10.1038/ncomms6274      0  \n",
       "2          NaN  10.1111/j.0953-816x.2004.03164.x      0  \n",
       "3          NaN         10.1142/s0219720009004114      1  \n",
       "4   NIHMS74438         10.1016/j.sbi.2008.04.004      0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 184 entries, 0 to 183\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              184 non-null    int64 \n",
      " 1   Title             184 non-null    object\n",
      " 2   Authors           184 non-null    object\n",
      " 3   Citation          184 non-null    object\n",
      " 4   First Author      184 non-null    object\n",
      " 5   Journal/Book      184 non-null    object\n",
      " 6   Publication Year  184 non-null    int64 \n",
      " 7   Create Date       184 non-null    object\n",
      " 8   PMCID             108 non-null    object\n",
      " 9   NIHMS ID          23 non-null     object\n",
      " 10  DOI               184 non-null    object\n",
      " 11  Label             184 non-null    int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 17.4+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27249342</td>\n",
       "      <td>recent advances in dynamic m6a rna modification</td>\n",
       "      <td>Cao G, Li HB, Yin Z, Flavell RA.</td>\n",
       "      <td>Open Biol. 2016 Apr;6(4):160003. doi: 10.1098/...</td>\n",
       "      <td>Cao G</td>\n",
       "      <td>Open Biol</td>\n",
       "      <td>2016</td>\n",
       "      <td>2016/06/02</td>\n",
       "      <td>PMC4852458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1098/rsob.160003</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16420705</td>\n",
       "      <td>discovery of time-delayed gene regulatory netw...</td>\n",
       "      <td>Li X, Rao S, Jiang W, Li C, Xiao Y, Guo Z, Zha...</td>\n",
       "      <td>BMC Bioinformatics. 2006 Jan 18;7:26. doi: 10....</td>\n",
       "      <td>Li X</td>\n",
       "      <td>BMC Bioinformatics</td>\n",
       "      <td>2006</td>\n",
       "      <td>2006/01/20</td>\n",
       "      <td>PMC1386718</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/1471-2105-7-26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27409645</td>\n",
       "      <td>evaluation of immortalized avpv- and arcuate-s...</td>\n",
       "      <td>Jacobs DC, Veitch RE, Chappell PE.</td>\n",
       "      <td>Endocrinology. 2016 Sep;157(9):3410-9. doi: 10...</td>\n",
       "      <td>Jacobs DC</td>\n",
       "      <td>Endocrinology</td>\n",
       "      <td>2016</td>\n",
       "      <td>2016/07/14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1210/en.2016-1294</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29615554</td>\n",
       "      <td>flying the rna nest: drosophila reveals novel ...</td>\n",
       "      <td>Lefebvre FA, Lécuyer É.</td>\n",
       "      <td>J Dev Biol. 2018 Mar 7;6(1):5. doi: 10.3390/jd...</td>\n",
       "      <td>Lefebvre FA</td>\n",
       "      <td>J Dev Biol</td>\n",
       "      <td>2018</td>\n",
       "      <td>2018/04/05</td>\n",
       "      <td>PMC5875563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.3390/jdb6010005</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23157550</td>\n",
       "      <td>exploring the dynamics of four rna-dependent r...</td>\n",
       "      <td>Shen H, Moustafa IM, Cameron CE, Colina CM.</td>\n",
       "      <td>J Phys Chem B. 2012 Dec 20;116(50):14515-24. d...</td>\n",
       "      <td>Shen H</td>\n",
       "      <td>J Phys Chem B</td>\n",
       "      <td>2012</td>\n",
       "      <td>2012/11/20</td>\n",
       "      <td>PMC5718632</td>\n",
       "      <td>NIHMS922948</td>\n",
       "      <td>10.1021/jp302709v</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  27249342    recent advances in dynamic m6a rna modification   \n",
       "1  16420705  discovery of time-delayed gene regulatory netw...   \n",
       "2  27409645  evaluation of immortalized avpv- and arcuate-s...   \n",
       "3  29615554  flying the rna nest: drosophila reveals novel ...   \n",
       "4  23157550  exploring the dynamics of four rna-dependent r...   \n",
       "\n",
       "                                             Authors  \\\n",
       "0                   Cao G, Li HB, Yin Z, Flavell RA.   \n",
       "1  Li X, Rao S, Jiang W, Li C, Xiao Y, Guo Z, Zha...   \n",
       "2                 Jacobs DC, Veitch RE, Chappell PE.   \n",
       "3                            Lefebvre FA, Lécuyer É.   \n",
       "4        Shen H, Moustafa IM, Cameron CE, Colina CM.   \n",
       "\n",
       "                                            Citation First Author  \\\n",
       "0  Open Biol. 2016 Apr;6(4):160003. doi: 10.1098/...        Cao G   \n",
       "1  BMC Bioinformatics. 2006 Jan 18;7:26. doi: 10....         Li X   \n",
       "2  Endocrinology. 2016 Sep;157(9):3410-9. doi: 10...    Jacobs DC   \n",
       "3  J Dev Biol. 2018 Mar 7;6(1):5. doi: 10.3390/jd...  Lefebvre FA   \n",
       "4  J Phys Chem B. 2012 Dec 20;116(50):14515-24. d...       Shen H   \n",
       "\n",
       "         Journal/Book  Publication Year Create Date       PMCID     NIHMS ID  \\\n",
       "0           Open Biol              2016  2016/06/02  PMC4852458          NaN   \n",
       "1  BMC Bioinformatics              2006  2006/01/20  PMC1386718          NaN   \n",
       "2       Endocrinology              2016  2016/07/14         NaN          NaN   \n",
       "3          J Dev Biol              2018  2018/04/05  PMC5875563          NaN   \n",
       "4       J Phys Chem B              2012  2012/11/20  PMC5718632  NIHMS922948   \n",
       "\n",
       "                      DOI  Label  \n",
       "0     10.1098/rsob.160003      0  \n",
       "1  10.1186/1471-2105-7-26      0  \n",
       "2    10.1210/en.2016-1294      1  \n",
       "3      10.3390/jdb6010005      1  \n",
       "4       10.1021/jp302709v      0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 460 entries, 0 to 459\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              460 non-null    int64 \n",
      " 1   Title             460 non-null    object\n",
      " 2   Authors           458 non-null    object\n",
      " 3   Citation          460 non-null    object\n",
      " 4   First Author      458 non-null    object\n",
      " 5   Journal/Book      460 non-null    object\n",
      " 6   Publication Year  460 non-null    int64 \n",
      " 7   Create Date       460 non-null    object\n",
      " 8   PMCID             250 non-null    object\n",
      " 9   NIHMS ID          51 non-null     object\n",
      " 10  DOI               459 non-null    object\n",
      " 11  Label             460 non-null    int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 43.2+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13677474</td>\n",
       "      <td>temporal progression of gene expression respon...</td>\n",
       "      <td>Wang H, Miyazaki S, Kawai K, Deyholos M, Galbr...</td>\n",
       "      <td>Plant Mol Biol. 2003 Jul;52(4):873-91. doi: 10...</td>\n",
       "      <td>Wang H</td>\n",
       "      <td>Plant Mol Biol</td>\n",
       "      <td>2003</td>\n",
       "      <td>2003/09/19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1023/a:1025029026375</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31997507</td>\n",
       "      <td>temporal profile of hev rna concentration in b...</td>\n",
       "      <td>Goel A, V Padmaprakash K, Benjamin M, Katiyar ...</td>\n",
       "      <td>J Viral Hepat. 2020 Jun;27(6):631-637. doi: 10...</td>\n",
       "      <td>Goel A</td>\n",
       "      <td>J Viral Hepat</td>\n",
       "      <td>2020</td>\n",
       "      <td>2020/01/31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1111/jvh.13266</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19147844</td>\n",
       "      <td>dynamic gene expression is required for anteri...</td>\n",
       "      <td>Pechmann M, McGregor AP, Schwager EE, Feitosa ...</td>\n",
       "      <td>Proc Natl Acad Sci U S A. 2009 Feb 3;106(5):14...</td>\n",
       "      <td>Pechmann M</td>\n",
       "      <td>Proc Natl Acad Sci U S A</td>\n",
       "      <td>2009</td>\n",
       "      <td>2009/01/17</td>\n",
       "      <td>PMC2635816</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1073/pnas.0811150106</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11606741</td>\n",
       "      <td>development of a two-part transcription probe ...</td>\n",
       "      <td>Li Z, Piggot PJ.</td>\n",
       "      <td>Proc Natl Acad Sci U S A. 2001 Oct 23;98(22):1...</td>\n",
       "      <td>Li Z</td>\n",
       "      <td>Proc Natl Acad Sci U S A</td>\n",
       "      <td>2001</td>\n",
       "      <td>2001/10/19</td>\n",
       "      <td>PMC60089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1073/pnas.221454798</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21176199</td>\n",
       "      <td>the complexity of gene expression dynamics rev...</td>\n",
       "      <td>Sun X, Zou Y, Nikiforova V, Kurths J, Walther D.</td>\n",
       "      <td>BMC Bioinformatics. 2010 Dec 22;11:607. doi: 1...</td>\n",
       "      <td>Sun X</td>\n",
       "      <td>BMC Bioinformatics</td>\n",
       "      <td>2010</td>\n",
       "      <td>2010/12/24</td>\n",
       "      <td>PMC3098107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/1471-2105-11-607</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  13677474  temporal progression of gene expression respon...   \n",
       "1  31997507  temporal profile of hev rna concentration in b...   \n",
       "2  19147844  dynamic gene expression is required for anteri...   \n",
       "3  11606741  development of a two-part transcription probe ...   \n",
       "4  21176199  the complexity of gene expression dynamics rev...   \n",
       "\n",
       "                                             Authors  \\\n",
       "0  Wang H, Miyazaki S, Kawai K, Deyholos M, Galbr...   \n",
       "1  Goel A, V Padmaprakash K, Benjamin M, Katiyar ...   \n",
       "2  Pechmann M, McGregor AP, Schwager EE, Feitosa ...   \n",
       "3                                   Li Z, Piggot PJ.   \n",
       "4   Sun X, Zou Y, Nikiforova V, Kurths J, Walther D.   \n",
       "\n",
       "                                            Citation First Author  \\\n",
       "0  Plant Mol Biol. 2003 Jul;52(4):873-91. doi: 10...       Wang H   \n",
       "1  J Viral Hepat. 2020 Jun;27(6):631-637. doi: 10...       Goel A   \n",
       "2  Proc Natl Acad Sci U S A. 2009 Feb 3;106(5):14...   Pechmann M   \n",
       "3  Proc Natl Acad Sci U S A. 2001 Oct 23;98(22):1...         Li Z   \n",
       "4  BMC Bioinformatics. 2010 Dec 22;11:607. doi: 1...        Sun X   \n",
       "\n",
       "               Journal/Book  Publication Year Create Date       PMCID  \\\n",
       "0            Plant Mol Biol              2003  2003/09/19         NaN   \n",
       "1             J Viral Hepat              2020  2020/01/31         NaN   \n",
       "2  Proc Natl Acad Sci U S A              2009  2009/01/17  PMC2635816   \n",
       "3  Proc Natl Acad Sci U S A              2001  2001/10/19    PMC60089   \n",
       "4        BMC Bioinformatics              2010  2010/12/24  PMC3098107   \n",
       "\n",
       "  NIHMS ID                       DOI  Label  \n",
       "0      NaN   10.1023/a:1025029026375      1  \n",
       "1      NaN         10.1111/jvh.13266      0  \n",
       "2      NaN   10.1073/pnas.0811150106      0  \n",
       "3      NaN    10.1073/pnas.221454798      1  \n",
       "4      NaN  10.1186/1471-2105-11-607      1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read data files.\n",
    "df_train = pd.read_csv(\"../data/rna_titles/rna_tuned_train.csv\")\n",
    "df_val = pd.read_csv(\"../data/rna_titles/rna_tuned_val.csv\")\n",
    "df_test = pd.read_csv(\"../data/rna_titles/rna_tuned_test.csv\")\n",
    "\n",
    "# Explore data.\n",
    "print(df_train.info())\n",
    "display(df_train.head())\n",
    "\n",
    "print(df_val.info())\n",
    "display(df_val.head())\n",
    "\n",
    "print(df_test.info())\n",
    "display(df_test.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e0b0f43-1abb-43d7-a914-6b86b0d38e50",
   "metadata": {},
   "source": [
    "## Process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "18c27fd9-692c-4b39-916c-e8f65ba3ae58",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1656, 4468)\n",
      "  (0, 4113)\t1\n",
      "  (0, 3680)\t1\n",
      "  (0, 713)\t2\n",
      "  (0, 2810)\t2\n",
      "  (0, 1646)\t1\n",
      "  (0, 1456)\t1\n",
      "  (0, 1974)\t1\n",
      "  (0, 2118)\t1\n",
      "  (0, 222)\t1\n",
      "  (0, 534)\t1\n",
      "  (0, 1511)\t1\n",
      "  (0, 209)\t1\n",
      "  (0, 267)\t1\n",
      "  (0, 1488)\t1\n",
      "  (1, 2810)\t1\n",
      "  (1, 1233)\t1\n",
      "  (1, 212)\t1\n",
      "  (1, 185)\t1\n",
      "  (1, 3136)\t1\n",
      "  (1, 1606)\t1\n",
      "  (1, 3553)\t1\n",
      "  (1, 3674)\t1\n",
      "  (1, 3519)\t1\n",
      "  (1, 4320)\t1\n",
      "  (1, 2213)\t1\n",
      "  :\t:\n",
      "  (1654, 4208)\t1\n",
      "  (1654, 99)\t1\n",
      "  (1654, 4356)\t1\n",
      "  (1654, 3442)\t1\n",
      "  (1654, 770)\t1\n",
      "  (1654, 3725)\t1\n",
      "  (1654, 1529)\t1\n",
      "  (1654, 858)\t1\n",
      "  (1654, 3763)\t1\n",
      "  (1654, 4460)\t1\n",
      "  (1654, 4395)\t1\n",
      "  (1655, 2810)\t1\n",
      "  (1655, 1646)\t1\n",
      "  (1655, 1456)\t1\n",
      "  (1655, 222)\t2\n",
      "  (1655, 1233)\t1\n",
      "  (1655, 4083)\t1\n",
      "  (1655, 2825)\t1\n",
      "  (1655, 3909)\t1\n",
      "  (1655, 3562)\t1\n",
      "  (1655, 1586)\t1\n",
      "  (1655, 2398)\t1\n",
      "  (1655, 665)\t1\n",
      "  (1655, 2305)\t1\n",
      "  (1655, 1897)\t1\n"
     ]
    }
   ],
   "source": [
    "# Build dictionary of features and transform documents to feature vectors.\n",
    "# Value of a word in the vocabulary is its frequency in the whole training corpus.\n",
    "# https://scikit-learn.org/stable/tutorial/text_analytics/working_with_text_data.html\n",
    "count_vect = CountVectorizer()\n",
    "X_train_counts = count_vect.fit_transform(df_train[\"Title\"])\n",
    "print(X_train_counts.shape)\n",
    "print(X_train_counts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9fe53584-c45a-4e1d-ba4f-ae59adb65087",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "'''\n",
    "# Regularize via “Term Frequency.”\n",
    "tf_transformer = TfidfTransformer(use_idf = False).fit(X_train_counts)\n",
    "X_train_tf = tf_transformer.transform(X_train_counts)\n",
    "print(X_train_tf.shape)\n",
    "print(X_train_tf)\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c73ebfce-996a-4e78-8ce7-8d3e61809ba6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1656, 4468)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Regularize via “Term Frequency times Inverse Document Frequency.”\n",
    "tfidf_transformer = TfidfTransformer()\n",
    "X_train_tfidf = tfidf_transformer.fit_transform(X_train_counts)\n",
    "X_train_tfidf.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "de2b8ba4-51a6-4032-b57f-38c922e7c3de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(460, 4468)\n",
      "  (0, 4129)\t0.18762593697179625\n",
      "  (0, 4055)\t0.13795009633576574\n",
      "  (0, 3704)\t0.4038868013919094\n",
      "  (0, 3590)\t0.45828339022661657\n",
      "  (0, 3566)\t0.4038868013919094\n",
      "  (0, 3501)\t0.31592974777919386\n",
      "  (0, 3243)\t0.35707917870897676\n",
      "  (0, 2810)\t0.08483258781007404\n",
      "  (0, 2359)\t0.38391176369580393\n",
      "  (0, 1974)\t0.10326528055314137\n",
      "  (0, 1646)\t0.0929566474021971\n",
      "  (0, 1456)\t0.09103243454822718\n",
      "  (1, 4427)\t0.33437719074511624\n",
      "  (1, 4055)\t0.10814230677478488\n",
      "  (1, 3889)\t0.35925906753300363\n",
      "  (1, 3553)\t0.08751230066393002\n",
      "  (1, 3408)\t0.3403893464400544\n",
      "  (1, 3229)\t0.23323058572619954\n",
      "  (1, 2979)\t0.24766471797497455\n",
      "  (1, 2810)\t0.06650224957528644\n",
      "  (1, 1974)\t0.161904137008808\n",
      "  (1, 1806)\t0.29474306498690495\n",
      "  (1, 1665)\t0.35925906753300363\n",
      "  (1, 1606)\t0.1615406481695132\n",
      "  (1, 799)\t0.3270010662599543\n",
      "  :\t:\n",
      "  (458, 4083)\t0.13031835678320042\n",
      "  (458, 4055)\t0.1339907359530225\n",
      "  (458, 3719)\t0.3371978032328088\n",
      "  (458, 2810)\t0.0823977740892973\n",
      "  (458, 2702)\t0.445130017032037\n",
      "  (458, 2575)\t0.23849646515784095\n",
      "  (458, 2560)\t0.3184332347692179\n",
      "  (458, 2558)\t0.29505322844349696\n",
      "  (458, 2076)\t0.42175001070631596\n",
      "  (458, 1646)\t0.09028866182760878\n",
      "  (458, 1456)\t0.08841967656930091\n",
      "  (458, 1264)\t0.2994909706625309\n",
      "  (458, 1233)\t0.15997852049761477\n",
      "  (458, 532)\t0.1886188180909217\n",
      "  (458, 222)\t0.10893550906589342\n",
      "  (459, 4083)\t0.15274740625674288\n",
      "  (459, 3562)\t0.36282018271830485\n",
      "  (459, 3553)\t0.12709149660236657\n",
      "  (459, 2562)\t0.4474899358178572\n",
      "  (459, 2140)\t0.34838547575820483\n",
      "  (459, 1305)\t0.4943372661712993\n",
      "  (459, 1233)\t0.18751237098127652\n",
      "  (459, 303)\t0.35967606181351436\n",
      "  (459, 222)\t0.12768436365994879\n",
      "  (459, 209)\t0.2864566735935201\n"
     ]
    }
   ],
   "source": [
    "# Call transform only, as featurizers have already been fit to training data.\n",
    "X_test_counts = count_vect.transform(df_test[\"Title\"])\n",
    "X_test_tfidf = tfidf_transformer.transform(X_test_counts)\n",
    "print(X_test_tfidf.shape)\n",
    "print(X_test_tfidf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c03a2300-ad05-447b-ad5c-a1b7b933e296",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1656\n",
      "0    0.620773\n",
      "1    0.379227\n",
      "Name: Label, dtype: float64\n",
      "0    1\n",
      "1    0\n",
      "2    0\n",
      "3    1\n",
      "4    0\n",
      "Name: Label, dtype: int64\n",
      "460\n",
      "0    0.621739\n",
      "1    0.378261\n",
      "Name: Label, dtype: float64\n",
      "0    1\n",
      "1    0\n",
      "2    0\n",
      "3    1\n",
      "4    1\n",
      "Name: Label, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Extract labels.\n",
    "y_train = df_train[\"Label\"]\n",
    "y_test = df_test[\"Label\"]\n",
    "\n",
    "print(len(y_train))\n",
    "print(y_train.value_counts(normalize = True))\n",
    "print(y_train.head())\n",
    "\n",
    "print(len(y_test))\n",
    "print(y_test.value_counts(normalize = True))\n",
    "print(y_test.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b974f10-bbd9-4d73-8de7-8d351cd04613",
   "metadata": {},
   "source": [
    "## Train"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4a2c37bf-5383-4fcc-a591-7a1594e57c01",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train a multinomial Naive Bayes model.\n",
    "mnb = MultinomialNB().fit(X_train_tfidf, y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "67ac471c-c794-4d93-a73b-2b9854c7cbd7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Train a linear SVM.\n",
    "# Regularized linear models with stochastic gradient descent (SGD) learning.\n",
    "svm = SGDClassifier(loss = \"hinge\", \n",
    "                    penalty = \"l2\",\n",
    "                    alpha = 1e-3, \n",
    "                    random_state = 42,\n",
    "                    max_iter = 5, \n",
    "                    tol = None)\n",
    "svm = svm.fit(X_train_tfidf, y_train)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e2e9e6a4-125e-4dd1-ad0a-4cdbccb11ee8",
   "metadata": {},
   "source": [
    "## Test"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a3ac8fa1-3714-4715-b4ee-92fb0d779c98",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- NAIVE BAYES ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.79      0.96      0.86       286\n",
      "           1       0.89      0.57      0.69       174\n",
      "\n",
      "    accuracy                           0.81       460\n",
      "   macro avg       0.84      0.76      0.78       460\n",
      "weighted avg       0.83      0.81      0.80       460\n",
      "\n",
      "\n",
      "--- NAIVE BAYES ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [274  12  75  99]\n",
      "F1             = 0.6947368421052632\n",
      "Accuracy       = 0.8108695652173913\n",
      "Precision      = 0.8918918918918919\n",
      "Recall         = 0.5689655172413793\n",
      "ROC AUC        = 0.7635037376416688\n",
      "---------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Predict on test set.\n",
    "y_mnb = mnb.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- NAIVE BAYES ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_mnb))\n",
    "\n",
    "# MNB performance metrics.\n",
    "# Care about most about false negatives – discarding papers we want to keep.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_mnb = metrics.confusion_matrix(y_test, y_mnb)\n",
    "acc_mnb = metrics.accuracy_score(y_test, y_mnb)\n",
    "f1_mnb = metrics.f1_score(y_test, y_mnb, zero_division = 0)\n",
    "precision_mnb = metrics.precision_score(y_test, y_mnb, zero_division = 0)\n",
    "recall_mnb = metrics.recall_score(y_test, y_mnb, zero_division = 0)\n",
    "roc_mnb = metrics.roc_auc_score(y_test, y_mnb)\n",
    "        \n",
    "print(\"\\n--- NAIVE BAYES ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_mnb.ravel())\n",
    "print(\"F1             =\", f1_mnb)\n",
    "print(\"Accuracy       =\", acc_mnb)\n",
    "print(\"Precision      =\", precision_mnb)\n",
    "print(\"Recall         =\", recall_mnb)\n",
    "print(\"ROC AUC        =\", roc_mnb)\n",
    "print(\"---------------------------------------------\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e0a972f0-0b2c-4f93-9c77-45983b11d066",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.87      0.86      0.87       286\n",
      "           1       0.78      0.79      0.78       174\n",
      "\n",
      "    accuracy                           0.83       460\n",
      "   macro avg       0.82      0.83      0.83       460\n",
      "weighted avg       0.84      0.83      0.84       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [246  40  36 138]\n",
      "F1             = 0.7840909090909092\n",
      "Accuracy       = 0.8347826086956521\n",
      "Precision      = 0.7752808988764045\n",
      "Recall         = 0.7931034482758621\n",
      "ROC AUC        = 0.8266216542078612\n",
      "---------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Predict on test set.\n",
    "y_svm = svm.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care about most about false negatives – discarding papers we want to keep.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm = metrics.confusion_matrix(y_test, y_svm)\n",
    "acc_svm = metrics.accuracy_score(y_test, y_svm)\n",
    "f1_svm = metrics.f1_score(y_test, y_svm, zero_division = 0)\n",
    "precision_svm = metrics.precision_score(y_test, y_svm, zero_division = 0)\n",
    "recall_svm = metrics.recall_score(y_test, y_svm, zero_division = 0)\n",
    "roc_svm = metrics.roc_auc_score(y_test, y_svm)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm.ravel())\n",
    "print(\"F1             =\", f1_svm)\n",
    "print(\"Accuracy       =\", acc_svm)\n",
    "print(\"Precision      =\", precision_svm)\n",
    "print(\"Recall         =\", recall_svm)\n",
    "print(\"ROC AUC        =\", roc_svm)\n",
    "print(\"---------------------------------------------\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e7ee1a8-c6f7-4ed6-8d28-14fd9ffcfc4f",
   "metadata": {},
   "source": [
    "## Hyperparameter tuning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b3f555dc-96a8-4422-bae3-636e949ffdb9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- TUNED NAIVE BAYES ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.81      0.94      0.87       286\n",
      "           1       0.86      0.64      0.74       174\n",
      "\n",
      "    accuracy                           0.83       460\n",
      "   macro avg       0.84      0.79      0.80       460\n",
      "weighted avg       0.83      0.83      0.82       460\n",
      "\n",
      "\n",
      "--- TUNED NAIVE BAYES ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [268  18  62 112]\n",
      "F1             = 0.7368421052631579\n",
      "Accuracy       = 0.8260869565217391\n",
      "Precision      = 0.8615384615384616\n",
      "Recall         = 0.6436781609195402\n",
      "ROC AUC        = 0.7903705489912387\n",
      "---------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Init search space.\n",
    "parameters = {'alpha': [0.0, 0.01, 0.1, 0.5, 1.0, 10.0]}\n",
    "\n",
    "# Tune SVM with 5-fold cross-validation.\n",
    "mnb_tuned = GridSearchCV(MultinomialNB(), \n",
    "                         parameters,\n",
    "                         cv = 5, \n",
    "                         n_jobs = -1)\n",
    "\n",
    "# Fit tuned model.\n",
    "mnb_tuned = mnb_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_mnb_tuned = mnb_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- TUNED NAIVE BAYES ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_mnb_tuned))\n",
    "\n",
    "# MNB tuned performance metrics.\n",
    "# Care most about false negatives.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_mnb_tuned = metrics.confusion_matrix(y_test, y_mnb_tuned)\n",
    "acc_mnb_tuned = metrics.accuracy_score(y_test, y_mnb_tuned)\n",
    "f1_mnb_tuned = metrics.f1_score(y_test, y_mnb_tuned, zero_division = 0)\n",
    "precision_mnb_tuned = metrics.precision_score(y_test, y_mnb_tuned, zero_division = 0)\n",
    "recall_mnb_tuned = metrics.recall_score(y_test, y_mnb_tuned, zero_division = 0)\n",
    "roc_mnb_tuned = metrics.roc_auc_score(y_test, y_mnb_tuned)\n",
    "        \n",
    "print(\"\\n--- TUNED NAIVE BAYES ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_mnb_tuned.ravel())\n",
    "print(\"F1             =\", f1_mnb_tuned)\n",
    "print(\"Accuracy       =\", acc_mnb_tuned)\n",
    "print(\"Precision      =\", precision_mnb_tuned)\n",
    "print(\"Recall         =\", recall_mnb_tuned)\n",
    "print(\"ROC AUC        =\", roc_mnb_tuned)\n",
    "print(\"---------------------------------------------\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0bc74e76-42c2-4b13-86e7-5d7dbea8a32c",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%capture\n",
    "'''\n",
    "# Init search space for SVM.\n",
    "parameters = {'clf__alpha': (1e-1, 1e-2, 1e-3)}\n",
    "\n",
    "# Tune SVM with 5-fold cross-validation.\n",
    "svm_tuned = SGDClassifier(loss = \"hinge\", \n",
    "                          penalty = \"l2\",\n",
    "                          random_state = 42,\n",
    "                          max_iter = 5, \n",
    "                          tol = None)\n",
    "svm_tuned = GridSearchCV(svm_tuned, \n",
    "                         parameters,\n",
    "                         cv = 5, \n",
    "                         n_jobs = -1)\n",
    "\n",
    "# Fit tuned model.\n",
    "svm_tuned = svm_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_tuned = svm_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- TUNED LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "becaafd9-f1dd-4fbd-b417-7776e4ff6bff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.68      0.99      0.81       286\n",
      "           1       0.93      0.23      0.37       174\n",
      "\n",
      "    accuracy                           0.70       460\n",
      "   macro avg       0.80      0.61      0.59       460\n",
      "weighted avg       0.77      0.70      0.64       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [283   3 134  40]\n",
      "F1             = 0.36866359447004604\n",
      "Accuracy       = 0.7021739130434783\n",
      "Precision      = 0.9302325581395349\n",
      "Recall         = 0.22988505747126436\n",
      "ROC AUC        = 0.609697773490877\n",
      "---------------------------------------------\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Train a linear SVM.\n",
    "# Regularized linear models with stochastic gradient descent (SGD) learning.\n",
    "svm_tuned = SGDClassifier(loss = \"hinge\", \n",
    "                          penalty = \"l2\",\n",
    "                          alpha = 1e-2, \n",
    "                          random_state = 42,\n",
    "                          max_iter = 5, \n",
    "                          tol = None)\n",
    "svm_tuned = svm_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_tuned = svm_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care most about false negatives.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm_tuned = metrics.confusion_matrix(y_test, y_svm_tuned)\n",
    "acc_svm_tuned = metrics.accuracy_score(y_test, y_svm_tuned)\n",
    "f1_svm_tuned = metrics.f1_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "precision_svm_tuned = metrics.precision_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "recall_svm_tuned = metrics.recall_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "roc_svm_tuned = metrics.roc_auc_score(y_test, y_svm_tuned)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm_tuned.ravel())\n",
    "print(\"F1             =\", f1_svm_tuned)\n",
    "print(\"Accuracy       =\", acc_svm_tuned)\n",
    "print(\"Precision      =\", precision_svm_tuned)\n",
    "print(\"Recall         =\", recall_svm_tuned)\n",
    "print(\"ROC AUC        =\", roc_svm_tuned)\n",
    "print(\"---------------------------------------------\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "705788cf-45f8-4474-aa90-a8d6e3a5e766",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.62      1.00      0.77       286\n",
      "           1       0.00      0.00      0.00       174\n",
      "\n",
      "    accuracy                           0.62       460\n",
      "   macro avg       0.31      0.50      0.38       460\n",
      "weighted avg       0.39      0.62      0.48       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [286   0 174   0]\n",
      "F1             = 0.0\n",
      "Accuracy       = 0.6217391304347826\n",
      "Precision      = 0.0\n",
      "Recall         = 0.0\n",
      "ROC AUC        = 0.5\n",
      "---------------------------------------------\n",
      "\n",
      "\n",
      "--- CLASSIFICATION REPORT ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.62      1.00      0.77       286\n",
      "           1       0.00      0.00      0.00       174\n",
      "\n",
      "    accuracy                           0.62       460\n",
      "   macro avg       0.31      0.50      0.38       460\n",
      "weighted avg       0.39      0.62      0.48       460\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/metrics/_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
      "  _warn_prf(average, modifier, msg_start, len(result))\n"
     ]
    }
   ],
   "source": [
    "# Train a linear SVM.\n",
    "# Regularized linear models with stochastic gradient descent (SGD) learning.\n",
    "svm_tuned = SGDClassifier(loss = \"hinge\", \n",
    "                    penalty = \"l2\",\n",
    "                    alpha = 1e-1, \n",
    "                    random_state = 42,\n",
    "                    max_iter = 5, \n",
    "                    tol = None)\n",
    "svm_tuned = svm_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_tuned = svm_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care most about false negatives.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm_tuned = metrics.confusion_matrix(y_test, y_svm_tuned)\n",
    "acc_svm_tuned = metrics.accuracy_score(y_test, y_svm_tuned)\n",
    "f1_svm_tuned = metrics.f1_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "precision_svm_tuned = metrics.precision_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "recall_svm_tuned = metrics.recall_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "roc_svm_tuned = metrics.roc_auc_score(y_test, y_svm_tuned)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm_tuned.ravel())\n",
    "print(\"F1             =\", f1_svm_tuned)\n",
    "print(\"Accuracy       =\", acc_svm_tuned)\n",
    "print(\"Precision      =\", precision_svm_tuned)\n",
    "print(\"Recall         =\", recall_svm_tuned)\n",
    "print(\"ROC AUC        =\", roc_svm_tuned)\n",
    "print(\"---------------------------------------------\\n\")\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- CLASSIFICATION REPORT ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "a4ee7ae7-2eb1-4228-984a-d8bdea073e32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.90      0.87      0.88       286\n",
      "           1       0.79      0.83      0.81       174\n",
      "\n",
      "    accuracy                           0.85       460\n",
      "   macro avg       0.84      0.85      0.85       460\n",
      "weighted avg       0.86      0.85      0.86       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [248  38  29 145]\n",
      "F1             = 0.8123249299719888\n",
      "Accuracy       = 0.8543478260869565\n",
      "Precision      = 0.7923497267759563\n",
      "Recall         = 0.8333333333333334\n",
      "ROC AUC        = 0.8502331002331003\n",
      "---------------------------------------------\n",
      "\n",
      "\n",
      "--- CLASSIFICATION REPORT ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.90      0.87      0.88       286\n",
      "           1       0.79      0.83      0.81       174\n",
      "\n",
      "    accuracy                           0.85       460\n",
      "   macro avg       0.84      0.85      0.85       460\n",
      "weighted avg       0.86      0.85      0.86       460\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Train a linear SVM.\n",
    "# Regularized linear models with stochastic gradient descent (SGD) learning.\n",
    "svm_tuned = SGDClassifier(loss = \"hinge\", \n",
    "                    penalty = \"l2\",\n",
    "                    alpha = 1e-4, \n",
    "                    random_state = 42,\n",
    "                    max_iter = 5, \n",
    "                    tol = None)\n",
    "svm_tuned = svm_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_tuned = svm_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care most about false negative.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm_tuned = metrics.confusion_matrix(y_test, y_svm_tuned)\n",
    "acc_svm_tuned = metrics.accuracy_score(y_test, y_svm_tuned)\n",
    "f1_svm_tuned = metrics.f1_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "precision_svm_tuned = metrics.precision_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "recall_svm_tuned = metrics.recall_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "roc_svm_tuned = metrics.roc_auc_score(y_test, y_svm_tuned)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm_tuned.ravel())\n",
    "print(\"F1             =\", f1_svm_tuned)\n",
    "print(\"Accuracy       =\", acc_svm_tuned)\n",
    "print(\"Precision      =\", precision_svm_tuned)\n",
    "print(\"Recall         =\", recall_svm_tuned)\n",
    "print(\"ROC AUC        =\", roc_svm_tuned)\n",
    "print(\"---------------------------------------------\\n\")\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- CLASSIFICATION REPORT ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "260113c7-254b-45ac-90f6-01c730937176",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.86      0.90      0.88       286\n",
      "           1       0.83      0.76      0.79       174\n",
      "\n",
      "    accuracy                           0.85       460\n",
      "   macro avg       0.84      0.83      0.84       460\n",
      "weighted avg       0.85      0.85      0.85       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [258  28  41 133]\n",
      "F1             = 0.7940298507462686\n",
      "Accuracy       = 0.85\n",
      "Precision      = 0.8260869565217391\n",
      "Recall         = 0.764367816091954\n",
      "ROC AUC        = 0.8332328590949281\n",
      "---------------------------------------------\n",
      "\n",
      "\n",
      "--- CLASSIFICATION REPORT ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.86      0.90      0.88       286\n",
      "           1       0.83      0.76      0.79       174\n",
      "\n",
      "    accuracy                           0.85       460\n",
      "   macro avg       0.84      0.83      0.84       460\n",
      "weighted avg       0.85      0.85      0.85       460\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Train a linear SVM.\n",
    "# Regularized linear models with stochastic gradient descent (SGD) learning.\n",
    "svm_tuned = SGDClassifier(loss = \"hinge\", \n",
    "                    penalty = \"l2\",\n",
    "                    alpha = 1e-5, \n",
    "                    random_state = 42,\n",
    "                    max_iter = 5, \n",
    "                    tol = None)\n",
    "svm_tuned = svm_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_tuned = svm_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care most about false negatives.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm_tuned = metrics.confusion_matrix(y_test, y_svm_tuned)\n",
    "acc_svm_tuned = metrics.accuracy_score(y_test, y_svm_tuned)\n",
    "f1_svm_tuned = metrics.f1_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "precision_svm_tuned = metrics.precision_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "recall_svm_tuned = metrics.recall_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "roc_svm_tuned = metrics.roc_auc_score(y_test, y_svm_tuned)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm_tuned.ravel())\n",
    "print(\"F1             =\", f1_svm_tuned)\n",
    "print(\"Accuracy       =\", acc_svm_tuned)\n",
    "print(\"Precision      =\", precision_svm_tuned)\n",
    "print(\"Recall         =\", recall_svm_tuned)\n",
    "print(\"ROC AUC        =\", roc_svm_tuned)\n",
    "print(\"---------------------------------------------\\n\")\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- CLASSIFICATION REPORT ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "d8e348cb-7974-4d9d-ba03-9a0625157922",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.87      0.88      0.87       286\n",
      "           1       0.80      0.78      0.79       174\n",
      "\n",
      "    accuracy                           0.84       460\n",
      "   macro avg       0.83      0.83      0.83       460\n",
      "weighted avg       0.84      0.84      0.84       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [252  34  38 136]\n",
      "F1             = 0.7906976744186047\n",
      "Accuracy       = 0.8434782608695652\n",
      "Precision      = 0.8\n",
      "Recall         = 0.7816091954022989\n",
      "ROC AUC        = 0.83136403826059\n",
      "---------------------------------------------\n",
      "\n",
      "\n",
      "--- CLASSIFICATION REPORT ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.87      0.88      0.87       286\n",
      "           1       0.80      0.78      0.79       174\n",
      "\n",
      "    accuracy                           0.84       460\n",
      "   macro avg       0.83      0.83      0.83       460\n",
      "weighted avg       0.84      0.84      0.84       460\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# Train a linear SVM.\n",
    "# Regularized linear models with stochastic gradient descent (SGD) learning.\n",
    "svm_tuned = SGDClassifier(loss = \"hinge\", \n",
    "                    penalty = \"l2\",\n",
    "                    alpha = 1e-6, \n",
    "                    random_state = 42,\n",
    "                    max_iter = 5, \n",
    "                    tol = None)\n",
    "svm_tuned = svm_tuned.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_tuned = svm_tuned.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care most about false negatives.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm_tuned = metrics.confusion_matrix(y_test, y_svm_tuned)\n",
    "acc_svm_tuned = metrics.accuracy_score(y_test, y_svm_tuned)\n",
    "f1_svm_tuned = metrics.f1_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "precision_svm_tuned = metrics.precision_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "recall_svm_tuned = metrics.recall_score(y_test, y_svm_tuned, zero_division = 0)\n",
    "roc_svm_tuned = metrics.roc_auc_score(y_test, y_svm_tuned)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm_tuned.ravel())\n",
    "print(\"F1             =\", f1_svm_tuned)\n",
    "print(\"Accuracy       =\", acc_svm_tuned)\n",
    "print(\"Precision      =\", precision_svm_tuned)\n",
    "print(\"Recall         =\", recall_svm_tuned)\n",
    "print(\"ROC AUC        =\", roc_svm_tuned)\n",
    "print(\"---------------------------------------------\\n\")\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- CLASSIFICATION REPORT ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_tuned))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ee75f0a-1942-4e06-b732-74e6a3118108",
   "metadata": {},
   "source": [
    "## Optimal model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "4cd11162-9c1d-4417-8fe6-07ee58e906a9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "              precision    recall  f1-score   support\n",
      "\n",
      "           0       0.90      0.87      0.88       286\n",
      "           1       0.79      0.83      0.81       174\n",
      "\n",
      "    accuracy                           0.85       460\n",
      "   macro avg       0.84      0.85      0.85       460\n",
      "weighted avg       0.86      0.85      0.86       460\n",
      "\n",
      "\n",
      "--- LINEAR SUPPORT VECTOR MACHINE ---\n",
      "\n",
      "\n",
      "---------------------------------------------\n",
      "tn, fp, fn, tp = [248  38  29 145]\n",
      "F1             = 0.8123249299719888\n",
      "Accuracy       = 0.8543478260869565\n",
      "Precision      = 0.7923497267759563\n",
      "Recall         = 0.8333333333333334\n",
      "ROC AUC        = 0.8502331002331003\n",
      "---------------------------------------------\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10\n",
      "  warnings.warn('alpha too small will result in numeric errors, '\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10\n",
      "  warnings.warn('alpha too small will result in numeric errors, '\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10\n",
      "  warnings.warn('alpha too small will result in numeric errors, '\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10\n",
      "  warnings.warn('alpha too small will result in numeric errors, '\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/naive_bayes.py:508: UserWarning: alpha too small will result in numeric errors, setting alpha = 1.0e-10\n",
      "  warnings.warn('alpha too small will result in numeric errors, '\n"
     ]
    }
   ],
   "source": [
    "# Fit model.\n",
    "svm_opt = SGDClassifier(loss = \"hinge\", \n",
    "                        penalty = \"l2\",\n",
    "                        alpha = 1e-4, \n",
    "                        random_state = 42,\n",
    "                        max_iter = 5, \n",
    "                        tol = None)\n",
    "svm_opt = svm_opt.fit(X_train_tfidf, y_train)\n",
    "\n",
    "# Predict on test set.\n",
    "y_svm_opt = svm_opt.predict(X_test_tfidf)\n",
    "\n",
    "# Evaluate performance.\n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(metrics.classification_report(y_test, y_svm_opt))\n",
    "\n",
    "# SVM performance metrics.\n",
    "# Care about most about false negatives – discarding papers we want to keep.\n",
    "# Recall, F1, and accuracy rely on false negatives.\n",
    "confusion_svm_opt = metrics.confusion_matrix(y_test, y_svm_opt)\n",
    "acc_svm_opt = metrics.accuracy_score(y_test, y_svm_opt)\n",
    "f1_svm_opt = metrics.f1_score(y_test, y_svm_opt, zero_division = 0)\n",
    "precision_svm_opt = metrics.precision_score(y_test, y_svm_opt, zero_division = 0)\n",
    "recall_svm_opt = metrics.recall_score(y_test, y_svm_opt, zero_division = 0)\n",
    "roc_svm_opt = metrics.roc_auc_score(y_test, y_svm_opt)\n",
    "        \n",
    "print(\"\\n--- LINEAR SUPPORT VECTOR MACHINE ---\\n\")\n",
    "print(\"\\n---------------------------------------------\")\n",
    "print(\"tn, fp, fn, tp =\", confusion_svm_opt.ravel())\n",
    "print(\"F1             =\", f1_svm_opt)\n",
    "print(\"Accuracy       =\", acc_svm_opt)\n",
    "print(\"Precision      =\", precision_svm_opt)\n",
    "print(\"Recall         =\", recall_svm_opt)\n",
    "print(\"ROC AUC        =\", roc_svm_opt)\n",
    "print(\"---------------------------------------------\\n\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67b0310a-fdfd-4602-acf8-25691ecb42c5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7b7c7db7-1061-4ceb-8212-defc47502597",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
