{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ffcb4bb9-12de-4189-85a3-3320ccf9be85",
   "metadata": {},
   "source": [
    "# Data programming with Snorkel: RNA"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db7f85e0-0682-418b-b705-eae6121cb10b",
   "metadata": {},
   "source": [
    "## Preamble"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6e0a69cf-15ef-45ee-8861-bb873fce2904",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Importations.\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from snorkel.labeling import labeling_function\n",
    "from snorkel.labeling import PandasLFApplier\n",
    "from snorkel.labeling import LFAnalysis\n",
    "from snorkel.labeling.model import MajorityLabelVoter\n",
    "from snorkel.labeling.model import LabelModel\n",
    "from snorkel.analysis import get_label_buckets\n",
    "from sklearn.model_selection import train_test_split\n",
    "import nltk\n",
    "from nltk.corpus import stopwords\n",
    "import re\n",
    "import string # For punctuation.\n",
    "import os\n",
    "import itertools\n",
    "\n",
    "# Define random state seed to be used throughout.\n",
    "seed = 123"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cb3423e-f514-4449-8052-40bd906b59f5",
   "metadata": {},
   "source": [
    "## Define functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01e33788-aa21-48d6-a27f-f26177eb7a72",
   "metadata": {},
   "source": [
    "## Read data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e8994334-d5ec-450c-8232-844af870baab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 2300 entries, 0 to 2299\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              2300 non-null   int64 \n",
      " 1   Title             2300 non-null   object\n",
      " 2   Authors           2297 non-null   object\n",
      " 3   Citation          2300 non-null   object\n",
      " 4   First Author      2297 non-null   object\n",
      " 5   Journal/Book      2300 non-null   object\n",
      " 6   Publication Year  2300 non-null   int64 \n",
      " 7   Create Date       2300 non-null   object\n",
      " 8   PMCID             1301 non-null   object\n",
      " 9   NIHMS ID          223 non-null    object\n",
      " 10  DOI               2299 non-null   object\n",
      " 11  Label             2300 non-null   int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 215.8+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>33485387</td>\n",
       "      <td>time-course single-cell rna sequencing reveals...</td>\n",
       "      <td>Sun C, Wang H, Ma Q, Chen C, Yue J, Li B, Zhan...</td>\n",
       "      <td>Cell Biosci. 2021 Jan 23;11(1):24. doi: 10.118...</td>\n",
       "      <td>Sun C</td>\n",
       "      <td>Cell Biosci</td>\n",
       "      <td>2021</td>\n",
       "      <td>2021/01/24</td>\n",
       "      <td>PMC7824938</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/s13578-021-00541-4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31852401</td>\n",
       "      <td>longitudinal rna-seq analysis of the repeatabi...</td>\n",
       "      <td>Rondina MT, Voora D, Simon LM, Schwertz H, Har...</td>\n",
       "      <td>Circ Res. 2020 Feb 14;126(4):501-516. doi: 10....</td>\n",
       "      <td>Rondina MT</td>\n",
       "      <td>Circ Res</td>\n",
       "      <td>2020</td>\n",
       "      <td>2019/12/20</td>\n",
       "      <td>PMC7323475</td>\n",
       "      <td>NIHMS1549147</td>\n",
       "      <td>10.1161/CIRCRESAHA.119.315215</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>30527648</td>\n",
       "      <td>time series analysis of gene changes and proce...</td>\n",
       "      <td>Wu D, Zhou M, Li L, Leng X, Zhang Z, Wang N, S...</td>\n",
       "      <td>Burns. 2019 Mar;45(2):387-397. doi: 10.1016/j....</td>\n",
       "      <td>Wu D</td>\n",
       "      <td>Burns</td>\n",
       "      <td>2019</td>\n",
       "      <td>2018/12/12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1016/j.burns.2018.08.022</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>25116415</td>\n",
       "      <td>a time course-dependent metastatic gene expres...</td>\n",
       "      <td>Chen R, Zhang G, Zhou Y, Li N, Lin J.</td>\n",
       "      <td>Diagn Pathol. 2014 Aug 13;9:155. doi: 10.1186/...</td>\n",
       "      <td>Chen R</td>\n",
       "      <td>Diagn Pathol</td>\n",
       "      <td>2014</td>\n",
       "      <td>2014/08/14</td>\n",
       "      <td>PMC4149277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/s13000-014-0155-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>28370374</td>\n",
       "      <td>the gene expression and immunohistochemical ti...</td>\n",
       "      <td>Mose KF, Burton M, Thomassen M, Andersen F, Kr...</td>\n",
       "      <td>Exp Dermatol. 2017 Oct;26(10):926-933. doi: 10...</td>\n",
       "      <td>Mose KF</td>\n",
       "      <td>Exp Dermatol</td>\n",
       "      <td>2017</td>\n",
       "      <td>2017/04/04</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1111/exd.13345</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  33485387  time-course single-cell rna sequencing reveals...   \n",
       "1  31852401  longitudinal rna-seq analysis of the repeatabi...   \n",
       "2  30527648  time series analysis of gene changes and proce...   \n",
       "3  25116415  a time course-dependent metastatic gene expres...   \n",
       "4  28370374  the gene expression and immunohistochemical ti...   \n",
       "\n",
       "                                             Authors  \\\n",
       "0  Sun C, Wang H, Ma Q, Chen C, Yue J, Li B, Zhan...   \n",
       "1  Rondina MT, Voora D, Simon LM, Schwertz H, Har...   \n",
       "2  Wu D, Zhou M, Li L, Leng X, Zhang Z, Wang N, S...   \n",
       "3              Chen R, Zhang G, Zhou Y, Li N, Lin J.   \n",
       "4  Mose KF, Burton M, Thomassen M, Andersen F, Kr...   \n",
       "\n",
       "                                            Citation First Author  \\\n",
       "0  Cell Biosci. 2021 Jan 23;11(1):24. doi: 10.118...        Sun C   \n",
       "1  Circ Res. 2020 Feb 14;126(4):501-516. doi: 10....   Rondina MT   \n",
       "2  Burns. 2019 Mar;45(2):387-397. doi: 10.1016/j....         Wu D   \n",
       "3  Diagn Pathol. 2014 Aug 13;9:155. doi: 10.1186/...       Chen R   \n",
       "4  Exp Dermatol. 2017 Oct;26(10):926-933. doi: 10...      Mose KF   \n",
       "\n",
       "   Journal/Book  Publication Year Create Date       PMCID      NIHMS ID  \\\n",
       "0   Cell Biosci              2021  2021/01/24  PMC7824938           NaN   \n",
       "1      Circ Res              2020  2019/12/20  PMC7323475  NIHMS1549147   \n",
       "2         Burns              2019  2018/12/12         NaN           NaN   \n",
       "3  Diagn Pathol              2014  2014/08/14  PMC4149277           NaN   \n",
       "4  Exp Dermatol              2017  2017/04/04         NaN           NaN   \n",
       "\n",
       "                             DOI  Label  \n",
       "0     10.1186/s13578-021-00541-4      1  \n",
       "1  10.1161/CIRCRESAHA.119.315215      1  \n",
       "2    10.1016/j.burns.2018.08.022      1  \n",
       "3      10.1186/s13000-014-0155-2      1  \n",
       "4              10.1111/exd.13345      1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Read data files.\n",
    "df = pd.read_csv(\"../data/rna_titles/timecourse_rna_titles_2300.csv\")\n",
    "\n",
    "# Explore data.\n",
    "print(df.info())\n",
    "display(df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0465cffc-5c6b-4da3-bf7c-26f270e41074",
   "metadata": {},
   "source": [
    "## Process data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "40762c2d-21c1-43ff-bc40-fa0fc276e107",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1840 entries, 0 to 1839\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              1840 non-null   int64 \n",
      " 1   Title             1840 non-null   object\n",
      " 2   Authors           1839 non-null   object\n",
      " 3   Citation          1840 non-null   object\n",
      " 4   First Author      1839 non-null   object\n",
      " 5   Journal/Book      1840 non-null   object\n",
      " 6   Publication Year  1840 non-null   int64 \n",
      " 7   Create Date       1840 non-null   object\n",
      " 8   PMCID             1051 non-null   object\n",
      " 9   NIHMS ID          172 non-null    object\n",
      " 10  DOI               1840 non-null   object\n",
      " 11  Label             1840 non-null   int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 172.6+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>16632472</td>\n",
       "      <td>structural, biochemical, and dynamic character...</td>\n",
       "      <td>Kang X, Hu Y, Li Y, Guo X, Jiang X, Lai L, Xia...</td>\n",
       "      <td>J Biol Chem. 2006 Jun 30;281(26):18216-26. doi...</td>\n",
       "      <td>Kang X</td>\n",
       "      <td>J Biol Chem</td>\n",
       "      <td>2006</td>\n",
       "      <td>2006/04/25</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1074/jbc.M513241200</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>29149186</td>\n",
       "      <td>cell cycle time series gene expression data en...</td>\n",
       "      <td>Szedlak A, Sims S, Smith N, Paternostro G, Pie...</td>\n",
       "      <td>PLoS Comput Biol. 2017 Nov 17;13(11):e1005849....</td>\n",
       "      <td>Szedlak A</td>\n",
       "      <td>PLoS Comput Biol</td>\n",
       "      <td>2017</td>\n",
       "      <td>2017/11/18</td>\n",
       "      <td>PMC5711035</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1371/journal.pcbi.1005849</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15013133</td>\n",
       "      <td>temporal patterns of gene expression after acu...</td>\n",
       "      <td>Lee CW, Stabile E, Kinnaird T, Shou M, Devaney...</td>\n",
       "      <td>J Am Coll Cardiol. 2004 Feb 4;43(3):474-82. do...</td>\n",
       "      <td>Lee CW</td>\n",
       "      <td>J Am Coll Cardiol</td>\n",
       "      <td>2004</td>\n",
       "      <td>2004/03/12</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1016/j.jacc.2003.09.033</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>27071649</td>\n",
       "      <td>saxs studies of rna: structures, dynamics, and...</td>\n",
       "      <td>Chen Y, Pollack L.</td>\n",
       "      <td>Wiley Interdiscip Rev RNA. 2016 Jul;7(4):512-2...</td>\n",
       "      <td>Chen Y</td>\n",
       "      <td>Wiley Interdiscip Rev RNA</td>\n",
       "      <td>2016</td>\n",
       "      <td>2016/04/14</td>\n",
       "      <td>PMC4909577</td>\n",
       "      <td>NIHMS768254</td>\n",
       "      <td>10.1002/wrna.1349</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21894962</td>\n",
       "      <td>mapping the landscape of rna dynamics with nmr...</td>\n",
       "      <td>Rinnenthal J, Buck J, Ferner J, Wacker A, Fürt...</td>\n",
       "      <td>Acc Chem Res. 2011 Dec 20;44(12):1292-301. doi...</td>\n",
       "      <td>Rinnenthal J</td>\n",
       "      <td>Acc Chem Res</td>\n",
       "      <td>2011</td>\n",
       "      <td>2011/09/08</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1021/ar200137d</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  16632472  structural, biochemical, and dynamic character...   \n",
       "1  29149186  cell cycle time series gene expression data en...   \n",
       "2  15013133  temporal patterns of gene expression after acu...   \n",
       "3  27071649  saxs studies of rna: structures, dynamics, and...   \n",
       "4  21894962  mapping the landscape of rna dynamics with nmr...   \n",
       "\n",
       "                                             Authors  \\\n",
       "0  Kang X, Hu Y, Li Y, Guo X, Jiang X, Lai L, Xia...   \n",
       "1  Szedlak A, Sims S, Smith N, Paternostro G, Pie...   \n",
       "2  Lee CW, Stabile E, Kinnaird T, Shou M, Devaney...   \n",
       "3                                 Chen Y, Pollack L.   \n",
       "4  Rinnenthal J, Buck J, Ferner J, Wacker A, Fürt...   \n",
       "\n",
       "                                            Citation  First Author  \\\n",
       "0  J Biol Chem. 2006 Jun 30;281(26):18216-26. doi...        Kang X   \n",
       "1  PLoS Comput Biol. 2017 Nov 17;13(11):e1005849....     Szedlak A   \n",
       "2  J Am Coll Cardiol. 2004 Feb 4;43(3):474-82. do...        Lee CW   \n",
       "3  Wiley Interdiscip Rev RNA. 2016 Jul;7(4):512-2...        Chen Y   \n",
       "4  Acc Chem Res. 2011 Dec 20;44(12):1292-301. doi...  Rinnenthal J   \n",
       "\n",
       "                Journal/Book  Publication Year Create Date       PMCID  \\\n",
       "0                J Biol Chem              2006  2006/04/25         NaN   \n",
       "1           PLoS Comput Biol              2017  2017/11/18  PMC5711035   \n",
       "2          J Am Coll Cardiol              2004  2004/03/12         NaN   \n",
       "3  Wiley Interdiscip Rev RNA              2016  2016/04/14  PMC4909577   \n",
       "4               Acc Chem Res              2011  2011/09/08         NaN   \n",
       "\n",
       "      NIHMS ID                           DOI  Label  \n",
       "0          NaN        10.1074/jbc.M513241200      0  \n",
       "1          NaN  10.1371/journal.pcbi.1005849      1  \n",
       "2          NaN    10.1016/j.jacc.2003.09.033      1  \n",
       "3  NIHMS768254             10.1002/wrna.1349      0  \n",
       "4          NaN             10.1021/ar200137d      0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 460 entries, 0 to 459\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              460 non-null    int64 \n",
      " 1   Title             460 non-null    object\n",
      " 2   Authors           458 non-null    object\n",
      " 3   Citation          460 non-null    object\n",
      " 4   First Author      458 non-null    object\n",
      " 5   Journal/Book      460 non-null    object\n",
      " 6   Publication Year  460 non-null    int64 \n",
      " 7   Create Date       460 non-null    object\n",
      " 8   PMCID             250 non-null    object\n",
      " 9   NIHMS ID          51 non-null     object\n",
      " 10  DOI               459 non-null    object\n",
      " 11  Label             460 non-null    int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 43.2+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>13677474</td>\n",
       "      <td>temporal progression of gene expression respon...</td>\n",
       "      <td>Wang H, Miyazaki S, Kawai K, Deyholos M, Galbr...</td>\n",
       "      <td>Plant Mol Biol. 2003 Jul;52(4):873-91. doi: 10...</td>\n",
       "      <td>Wang H</td>\n",
       "      <td>Plant Mol Biol</td>\n",
       "      <td>2003</td>\n",
       "      <td>2003/09/19</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1023/a:1025029026375</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31997507</td>\n",
       "      <td>temporal profile of hev rna concentration in b...</td>\n",
       "      <td>Goel A, V Padmaprakash K, Benjamin M, Katiyar ...</td>\n",
       "      <td>J Viral Hepat. 2020 Jun;27(6):631-637. doi: 10...</td>\n",
       "      <td>Goel A</td>\n",
       "      <td>J Viral Hepat</td>\n",
       "      <td>2020</td>\n",
       "      <td>2020/01/31</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1111/jvh.13266</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19147844</td>\n",
       "      <td>dynamic gene expression is required for anteri...</td>\n",
       "      <td>Pechmann M, McGregor AP, Schwager EE, Feitosa ...</td>\n",
       "      <td>Proc Natl Acad Sci U S A. 2009 Feb 3;106(5):14...</td>\n",
       "      <td>Pechmann M</td>\n",
       "      <td>Proc Natl Acad Sci U S A</td>\n",
       "      <td>2009</td>\n",
       "      <td>2009/01/17</td>\n",
       "      <td>PMC2635816</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1073/pnas.0811150106</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>11606741</td>\n",
       "      <td>development of a two-part transcription probe ...</td>\n",
       "      <td>Li Z, Piggot PJ.</td>\n",
       "      <td>Proc Natl Acad Sci U S A. 2001 Oct 23;98(22):1...</td>\n",
       "      <td>Li Z</td>\n",
       "      <td>Proc Natl Acad Sci U S A</td>\n",
       "      <td>2001</td>\n",
       "      <td>2001/10/19</td>\n",
       "      <td>PMC60089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1073/pnas.221454798</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>21176199</td>\n",
       "      <td>the complexity of gene expression dynamics rev...</td>\n",
       "      <td>Sun X, Zou Y, Nikiforova V, Kurths J, Walther D.</td>\n",
       "      <td>BMC Bioinformatics. 2010 Dec 22;11:607. doi: 1...</td>\n",
       "      <td>Sun X</td>\n",
       "      <td>BMC Bioinformatics</td>\n",
       "      <td>2010</td>\n",
       "      <td>2010/12/24</td>\n",
       "      <td>PMC3098107</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/1471-2105-11-607</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  13677474  temporal progression of gene expression respon...   \n",
       "1  31997507  temporal profile of hev rna concentration in b...   \n",
       "2  19147844  dynamic gene expression is required for anteri...   \n",
       "3  11606741  development of a two-part transcription probe ...   \n",
       "4  21176199  the complexity of gene expression dynamics rev...   \n",
       "\n",
       "                                             Authors  \\\n",
       "0  Wang H, Miyazaki S, Kawai K, Deyholos M, Galbr...   \n",
       "1  Goel A, V Padmaprakash K, Benjamin M, Katiyar ...   \n",
       "2  Pechmann M, McGregor AP, Schwager EE, Feitosa ...   \n",
       "3                                   Li Z, Piggot PJ.   \n",
       "4   Sun X, Zou Y, Nikiforova V, Kurths J, Walther D.   \n",
       "\n",
       "                                            Citation First Author  \\\n",
       "0  Plant Mol Biol. 2003 Jul;52(4):873-91. doi: 10...       Wang H   \n",
       "1  J Viral Hepat. 2020 Jun;27(6):631-637. doi: 10...       Goel A   \n",
       "2  Proc Natl Acad Sci U S A. 2009 Feb 3;106(5):14...   Pechmann M   \n",
       "3  Proc Natl Acad Sci U S A. 2001 Oct 23;98(22):1...         Li Z   \n",
       "4  BMC Bioinformatics. 2010 Dec 22;11:607. doi: 1...        Sun X   \n",
       "\n",
       "               Journal/Book  Publication Year Create Date       PMCID  \\\n",
       "0            Plant Mol Biol              2003  2003/09/19         NaN   \n",
       "1             J Viral Hepat              2020  2020/01/31         NaN   \n",
       "2  Proc Natl Acad Sci U S A              2009  2009/01/17  PMC2635816   \n",
       "3  Proc Natl Acad Sci U S A              2001  2001/10/19    PMC60089   \n",
       "4        BMC Bioinformatics              2010  2010/12/24  PMC3098107   \n",
       "\n",
       "  NIHMS ID                       DOI  Label  \n",
       "0      NaN   10.1023/a:1025029026375      1  \n",
       "1      NaN         10.1111/jvh.13266      0  \n",
       "2      NaN   10.1073/pnas.0811150106      0  \n",
       "3      NaN    10.1073/pnas.221454798      1  \n",
       "4      NaN  10.1186/1471-2105-11-607      1  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Train-test split (80% / 20%, stratified by gold label value).\n",
    "X_train, X_test, y_train, y_test = train_test_split(df[\"Title\"], \n",
    "                                                    df[\"Label\"], \n",
    "                                                    test_size = 0.2, \n",
    "                                                    random_state = 42,\n",
    "                                                    stratify = df[\"Label\"])\n",
    "df_train = df.iloc[X_train.index].reset_index(drop = True)\n",
    "df_test = df.iloc[X_test.index].reset_index(drop = True)\n",
    "\n",
    "# Set data types.\n",
    "X_train = pd.DataFrame(X_train)\n",
    "X_test = pd.DataFrame(X_test)\n",
    "y_train = y_train.to_numpy(dtype = int)\n",
    "y_test = y_test.to_numpy(dtype = int)\n",
    "\n",
    "print(df_train.info())\n",
    "display(df_train.head())\n",
    "print(df_test.info())\n",
    "display(df_test.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f43e9840-8b1c-422e-8e07-5ae7b63385f5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 1656 entries, 0 to 1655\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              1656 non-null   int64 \n",
      " 1   Title             1656 non-null   object\n",
      " 2   Authors           1655 non-null   object\n",
      " 3   Citation          1656 non-null   object\n",
      " 4   First Author      1655 non-null   object\n",
      " 5   Journal/Book      1656 non-null   object\n",
      " 6   Publication Year  1656 non-null   int64 \n",
      " 7   Create Date       1656 non-null   object\n",
      " 8   PMCID             943 non-null    object\n",
      " 9   NIHMS ID          149 non-null    object\n",
      " 10  DOI               1656 non-null   object\n",
      " 11  Label             1656 non-null   int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 155.4+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>21205307</td>\n",
       "      <td>time-series clustering of gene expression in i...</td>\n",
       "      <td>Ghandhi SA, Sinha A, Markatou M, Amundson SA.</td>\n",
       "      <td>BMC Genomics. 2011 Jan 4;12:2. doi: 10.1186/14...</td>\n",
       "      <td>Ghandhi SA</td>\n",
       "      <td>BMC Genomics</td>\n",
       "      <td>2011</td>\n",
       "      <td>2011/01/06</td>\n",
       "      <td>PMC3022823</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/1471-2164-12-2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>25409906</td>\n",
       "      <td>dynamic analyses of alternative polyadenylatio...</td>\n",
       "      <td>Xia Z, Donehower LA, Cooper TA, Neilson JR, Wh...</td>\n",
       "      <td>Nat Commun. 2014 Nov 20;5:5274. doi: 10.1038/n...</td>\n",
       "      <td>Xia Z</td>\n",
       "      <td>Nat Commun</td>\n",
       "      <td>2014</td>\n",
       "      <td>2014/11/21</td>\n",
       "      <td>PMC4467577</td>\n",
       "      <td>NIHMS674338</td>\n",
       "      <td>10.1038/ncomms6274</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14984408</td>\n",
       "      <td>temporal and spatial patterns of kv1.1 and kv1...</td>\n",
       "      <td>Karimi-Abdolrezaee S, Eftekharpour E, Fehlings...</td>\n",
       "      <td>Eur J Neurosci. 2004 Feb;19(3):577-89. doi: 10...</td>\n",
       "      <td>Karimi-Abdolrezaee S</td>\n",
       "      <td>Eur J Neurosci</td>\n",
       "      <td>2004</td>\n",
       "      <td>2004/02/27</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1111/j.0953-816x.2004.03164.x</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19340919</td>\n",
       "      <td>simultaneously segmenting multiple gene expres...</td>\n",
       "      <td>Tadepalli S, Ramakrishnan N, Watson LT, Mishra...</td>\n",
       "      <td>J Bioinform Comput Biol. 2009 Apr;7(2):339-56....</td>\n",
       "      <td>Tadepalli S</td>\n",
       "      <td>J Bioinform Comput Biol</td>\n",
       "      <td>2009</td>\n",
       "      <td>2009/04/03</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1142/s0219720009004114</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>18547802</td>\n",
       "      <td>rna dynamics: it is about time</td>\n",
       "      <td>Al-Hashimi HM, Walter NG.</td>\n",
       "      <td>Curr Opin Struct Biol. 2008 Jun;18(3):321-9. d...</td>\n",
       "      <td>Al-Hashimi HM</td>\n",
       "      <td>Curr Opin Struct Biol</td>\n",
       "      <td>2008</td>\n",
       "      <td>2008/06/13</td>\n",
       "      <td>PMC2580758</td>\n",
       "      <td>NIHMS74438</td>\n",
       "      <td>10.1016/j.sbi.2008.04.004</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  21205307  time-series clustering of gene expression in i...   \n",
       "1  25409906  dynamic analyses of alternative polyadenylatio...   \n",
       "2  14984408  temporal and spatial patterns of kv1.1 and kv1...   \n",
       "3  19340919  simultaneously segmenting multiple gene expres...   \n",
       "4  18547802                     rna dynamics: it is about time   \n",
       "\n",
       "                                             Authors  \\\n",
       "0      Ghandhi SA, Sinha A, Markatou M, Amundson SA.   \n",
       "1  Xia Z, Donehower LA, Cooper TA, Neilson JR, Wh...   \n",
       "2  Karimi-Abdolrezaee S, Eftekharpour E, Fehlings...   \n",
       "3  Tadepalli S, Ramakrishnan N, Watson LT, Mishra...   \n",
       "4                          Al-Hashimi HM, Walter NG.   \n",
       "\n",
       "                                            Citation          First Author  \\\n",
       "0  BMC Genomics. 2011 Jan 4;12:2. doi: 10.1186/14...            Ghandhi SA   \n",
       "1  Nat Commun. 2014 Nov 20;5:5274. doi: 10.1038/n...                 Xia Z   \n",
       "2  Eur J Neurosci. 2004 Feb;19(3):577-89. doi: 10...  Karimi-Abdolrezaee S   \n",
       "3  J Bioinform Comput Biol. 2009 Apr;7(2):339-56....           Tadepalli S   \n",
       "4  Curr Opin Struct Biol. 2008 Jun;18(3):321-9. d...         Al-Hashimi HM   \n",
       "\n",
       "              Journal/Book  Publication Year Create Date       PMCID  \\\n",
       "0             BMC Genomics              2011  2011/01/06  PMC3022823   \n",
       "1               Nat Commun              2014  2014/11/21  PMC4467577   \n",
       "2           Eur J Neurosci              2004  2004/02/27         NaN   \n",
       "3  J Bioinform Comput Biol              2009  2009/04/03         NaN   \n",
       "4    Curr Opin Struct Biol              2008  2008/06/13  PMC2580758   \n",
       "\n",
       "      NIHMS ID                               DOI  Label  \n",
       "0          NaN            10.1186/1471-2164-12-2      1  \n",
       "1  NIHMS674338                10.1038/ncomms6274      0  \n",
       "2          NaN  10.1111/j.0953-816x.2004.03164.x      0  \n",
       "3          NaN         10.1142/s0219720009004114      1  \n",
       "4   NIHMS74438         10.1016/j.sbi.2008.04.004      0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n",
      "RangeIndex: 184 entries, 0 to 183\n",
      "Data columns (total 12 columns):\n",
      " #   Column            Non-Null Count  Dtype \n",
      "---  ------            --------------  ----- \n",
      " 0   PMID              184 non-null    int64 \n",
      " 1   Title             184 non-null    object\n",
      " 2   Authors           184 non-null    object\n",
      " 3   Citation          184 non-null    object\n",
      " 4   First Author      184 non-null    object\n",
      " 5   Journal/Book      184 non-null    object\n",
      " 6   Publication Year  184 non-null    int64 \n",
      " 7   Create Date       184 non-null    object\n",
      " 8   PMCID             108 non-null    object\n",
      " 9   NIHMS ID          23 non-null     object\n",
      " 10  DOI               184 non-null    object\n",
      " 11  Label             184 non-null    int64 \n",
      "dtypes: int64(3), object(9)\n",
      "memory usage: 17.4+ KB\n",
      "None\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PMID</th>\n",
       "      <th>Title</th>\n",
       "      <th>Authors</th>\n",
       "      <th>Citation</th>\n",
       "      <th>First Author</th>\n",
       "      <th>Journal/Book</th>\n",
       "      <th>Publication Year</th>\n",
       "      <th>Create Date</th>\n",
       "      <th>PMCID</th>\n",
       "      <th>NIHMS ID</th>\n",
       "      <th>DOI</th>\n",
       "      <th>Label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>27249342</td>\n",
       "      <td>recent advances in dynamic m6a rna modification</td>\n",
       "      <td>Cao G, Li HB, Yin Z, Flavell RA.</td>\n",
       "      <td>Open Biol. 2016 Apr;6(4):160003. doi: 10.1098/...</td>\n",
       "      <td>Cao G</td>\n",
       "      <td>Open Biol</td>\n",
       "      <td>2016</td>\n",
       "      <td>2016/06/02</td>\n",
       "      <td>PMC4852458</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1098/rsob.160003</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16420705</td>\n",
       "      <td>discovery of time-delayed gene regulatory netw...</td>\n",
       "      <td>Li X, Rao S, Jiang W, Li C, Xiao Y, Guo Z, Zha...</td>\n",
       "      <td>BMC Bioinformatics. 2006 Jan 18;7:26. doi: 10....</td>\n",
       "      <td>Li X</td>\n",
       "      <td>BMC Bioinformatics</td>\n",
       "      <td>2006</td>\n",
       "      <td>2006/01/20</td>\n",
       "      <td>PMC1386718</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1186/1471-2105-7-26</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>27409645</td>\n",
       "      <td>evaluation of immortalized avpv- and arcuate-s...</td>\n",
       "      <td>Jacobs DC, Veitch RE, Chappell PE.</td>\n",
       "      <td>Endocrinology. 2016 Sep;157(9):3410-9. doi: 10...</td>\n",
       "      <td>Jacobs DC</td>\n",
       "      <td>Endocrinology</td>\n",
       "      <td>2016</td>\n",
       "      <td>2016/07/14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.1210/en.2016-1294</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>29615554</td>\n",
       "      <td>flying the rna nest: drosophila reveals novel ...</td>\n",
       "      <td>Lefebvre FA, Lécuyer É.</td>\n",
       "      <td>J Dev Biol. 2018 Mar 7;6(1):5. doi: 10.3390/jd...</td>\n",
       "      <td>Lefebvre FA</td>\n",
       "      <td>J Dev Biol</td>\n",
       "      <td>2018</td>\n",
       "      <td>2018/04/05</td>\n",
       "      <td>PMC5875563</td>\n",
       "      <td>NaN</td>\n",
       "      <td>10.3390/jdb6010005</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>23157550</td>\n",
       "      <td>exploring the dynamics of four rna-dependent r...</td>\n",
       "      <td>Shen H, Moustafa IM, Cameron CE, Colina CM.</td>\n",
       "      <td>J Phys Chem B. 2012 Dec 20;116(50):14515-24. d...</td>\n",
       "      <td>Shen H</td>\n",
       "      <td>J Phys Chem B</td>\n",
       "      <td>2012</td>\n",
       "      <td>2012/11/20</td>\n",
       "      <td>PMC5718632</td>\n",
       "      <td>NIHMS922948</td>\n",
       "      <td>10.1021/jp302709v</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       PMID                                              Title  \\\n",
       "0  27249342    recent advances in dynamic m6a rna modification   \n",
       "1  16420705  discovery of time-delayed gene regulatory netw...   \n",
       "2  27409645  evaluation of immortalized avpv- and arcuate-s...   \n",
       "3  29615554  flying the rna nest: drosophila reveals novel ...   \n",
       "4  23157550  exploring the dynamics of four rna-dependent r...   \n",
       "\n",
       "                                             Authors  \\\n",
       "0                   Cao G, Li HB, Yin Z, Flavell RA.   \n",
       "1  Li X, Rao S, Jiang W, Li C, Xiao Y, Guo Z, Zha...   \n",
       "2                 Jacobs DC, Veitch RE, Chappell PE.   \n",
       "3                            Lefebvre FA, Lécuyer É.   \n",
       "4        Shen H, Moustafa IM, Cameron CE, Colina CM.   \n",
       "\n",
       "                                            Citation First Author  \\\n",
       "0  Open Biol. 2016 Apr;6(4):160003. doi: 10.1098/...        Cao G   \n",
       "1  BMC Bioinformatics. 2006 Jan 18;7:26. doi: 10....         Li X   \n",
       "2  Endocrinology. 2016 Sep;157(9):3410-9. doi: 10...    Jacobs DC   \n",
       "3  J Dev Biol. 2018 Mar 7;6(1):5. doi: 10.3390/jd...  Lefebvre FA   \n",
       "4  J Phys Chem B. 2012 Dec 20;116(50):14515-24. d...       Shen H   \n",
       "\n",
       "         Journal/Book  Publication Year Create Date       PMCID     NIHMS ID  \\\n",
       "0           Open Biol              2016  2016/06/02  PMC4852458          NaN   \n",
       "1  BMC Bioinformatics              2006  2006/01/20  PMC1386718          NaN   \n",
       "2       Endocrinology              2016  2016/07/14         NaN          NaN   \n",
       "3          J Dev Biol              2018  2018/04/05  PMC5875563          NaN   \n",
       "4       J Phys Chem B              2012  2012/11/20  PMC5718632  NIHMS922948   \n",
       "\n",
       "                      DOI  Label  \n",
       "0     10.1098/rsob.160003      0  \n",
       "1  10.1186/1471-2105-7-26      0  \n",
       "2    10.1210/en.2016-1294      1  \n",
       "3      10.3390/jdb6010005      1  \n",
       "4       10.1021/jp302709v      0  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Split off validation set from training set.\n",
    "# Train-val split (90% / 10%, stratified by gold label value).\n",
    "X_train, X_val, y_train, y_val = train_test_split(X_train, \n",
    "                                                  y_train, \n",
    "                                                  test_size = 0.1, \n",
    "                                                  random_state = 42,\n",
    "                                                  stratify = y_train)\n",
    "df_train = df.iloc[X_train.index].reset_index(drop = True)\n",
    "df_val = df.iloc[X_val.index].reset_index(drop = True)\n",
    "\n",
    "# Set data types.\n",
    "X_train = pd.DataFrame(X_train)\n",
    "X_val = pd.DataFrame(X_val)\n",
    "\n",
    "print(df_train.info())\n",
    "display(df_train.head())\n",
    "print(df_val.info())\n",
    "display(df_val.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d929df13-9939-4229-90ed-3afbe550c30e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    0.62087\n",
      "1    0.37913\n",
      "Name: Label, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeMAAAD1CAYAAABjj50KAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAM4UlEQVR4nO3dX4ideX3H8c+3k+aiLcXSTNXmjwk1skRQsNPYgqX2YjGrhSgVmrVU+kdCCmnxomBu6o03Lt6UYmwIEkpvGgq1Nmg0F4K1YKXJynZp1sYOqZppLEYrylppjH57kbE9PTvZOZucyU/PvF4wcJ7f89tzvjfZN89zzpyp7g4AMM6PjB4AALY7MQaAwcQYAAYTYwAYTIwBYDAxBoDBdox64V27dvX+/ftHvTwAPHRPPvnkV7t7eXp9WIz379+fK1eujHp5AHjoquqLG627TQ0Ag4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYMO+9GPR7T/10dEj8AC+8N43jR4B2EZcGQPAYGIMAIOJMQAMNlOMq+pIVV2rqtWqOnWPPa+vqqeq6mpV/d18xwSAxbXpB7iqainJ6SSPJllLcrmqLnT3MxN7XpTkA0mOdPeXqupntmheAFg4s1wZH06y2t3Xu/t2kvNJjk7teVuSD3X3l5Kku78y3zEBYHHNEuPdSW5MHK+tr016RZKfqqpPVtWTVfX2jZ6oqo5X1ZWqunLr1q37mxgAFswsMa4N1nrqeEeSn0/ypiRvSPLHVfWK5/xH3We7e6W7V5aXl1/wsACwiGb50o+1JHsnjvckubnBnq9297eSfKuqPpXk1Uk+P5cpAWCBzXJlfDnJwao6UFU7kxxLcmFqz98m+eWq2lFVP5bktUk+N99RAWAxbXpl3N13qupkkktJlpKc6+6rVXVi/fyZ7v5cVX08ydNJvpfkg939z1s5OAAsipm+m7q7Lya5OLV2Zur4fUneN7/RAGB78A1cADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADDZTjKvqSFVdq6rVqjq1wfnXV9U3quqp9Z93z39UAFhMOzbbUFVLSU4neTTJWpLLVXWhu5+Z2vr33f1rWzAjACy0Wa6MDydZ7e7r3X07yfkkR7d2LADYPmaJ8e4kNyaO19bXpv1SVf1TVX2sql45l+kAYBvY9DZ1ktpgraeOP5vkZd39bFW9McmHkxx8zhNVHU9yPEn27dv3wiYFgAU1y5XxWpK9E8d7ktyc3NDd3+zuZ9cfX0zyo1W1a/qJuvtsd69098ry8vIDjA0Ai2OWGF9OcrCqDlTVziTHklyY3FBVL6mqWn98eP15vzbvYQFgEW16m7q771TVySSXkiwlOdfdV6vqxPr5M0nemuT3q+pOkm8nOdbd07eyAYANzPKe8fdvPV+cWjsz8fj9Sd4/39EAYHvwDVwAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMNlOMq+pIVV2rqtWqOvU8+36hqr5bVW+d34gAsNg2jXFVLSU5neSxJIeSPF5Vh+6x74kkl+Y9JAAsslmujA8nWe3u6919O8n5JEc32PcHSf46yVfmOB8ALLxZYrw7yY2J47X1tf9VVbuTvCXJmfmNBgDbwywxrg3Weur4T5K8q7u/+7xPVHW8qq5U1ZVbt27NOCIALLYdM+xZS7J34nhPkptTe1aSnK+qJNmV5I1Vdae7Pzy5qbvPJjmbJCsrK9NBB4BtaZYYX05ysKoOJPn3JMeSvG1yQ3cf+P7jqvrzJB+ZDjEAsLFNY9zdd6rqZO5+SnopybnuvlpVJ9bPe58YAB7ALFfG6e6LSS5OrW0Y4e7+7QcfCwC2D9/ABQCDiTEADDbTbWqAHxb7T3109Ag8gC+8902jRxjClTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAw2EwxrqojVXWtqlar6tQG549W1dNV9VRVXamq181/VABYTDs221BVS0lOJ3k0yVqSy1V1obufmdj2iSQXurur6lVJ/irJI1sxMAAsmlmujA8nWe3u6919O8n5JEcnN3T3s93d64c/nqQDAMxklhjvTnJj4nhtfe3/qaq3VNW/JPlokt/d6Imq6vj6bewrt27dup95AWDhzBLj2mDtOVe+3f033f1Ikjcnec9GT9TdZ7t7pbtXlpeXX9CgALCoZonxWpK9E8d7kty81+bu/lSSn6uqXQ84GwBsC7PE+HKSg1V1oKp2JjmW5MLkhqp6eVXV+uPXJNmZ5GvzHhYAFtGmn6bu7jtVdTLJpSRLSc5199WqOrF+/kySX0/y9qr6TpJvJ/mNiQ90AQDPY9MYJ0l3X0xycWrtzMTjJ5I8Md/RAGB78A1cADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADDZTjKvqSFVdq6rVqjq1wfnfrKqn138+XVWvnv+oALCYNo1xVS0lOZ3ksSSHkjxeVYemtv1bkl/p7lcleU+Ss/MeFAAW1SxXxoeTrHb39e6+neR8kqOTG7r709399fXDzyTZM98xAWBxzRLj3UluTByvra/dy+8l+diDDAUA28mOGfbUBmu94caqX83dGL/uHuePJzmeJPv27ZtxRABYbLNcGa8l2TtxvCfJzelNVfWqJB9McrS7v7bRE3X32e5e6e6V5eXl+5kXABbOLDG+nORgVR2oqp1JjiW5MLmhqvYl+VCS3+ruz89/TABYXJvepu7uO1V1MsmlJEtJznX31ao6sX7+TJJ3J/npJB+oqiS5090rWzc2ACyOWd4zTndfTHJxau3MxON3JHnHfEcDgO3BN3ABwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAw2EwxrqojVXWtqlar6tQG5x+pqn+oqv+uqj+a/5gAsLh2bLahqpaSnE7yaJK1JJer6kJ3PzOx7T+T/GGSN2/FkACwyGa5Mj6cZLW7r3f37STnkxyd3NDdX+nuy0m+swUzAsBCmyXGu5PcmDheW18DAOZglhjXBmt9Py9WVcer6kpVXbl169b9PAUALJxZYryWZO/E8Z4kN+/nxbr7bHevdPfK8vLy/TwFACycWWJ8OcnBqjpQVTuTHEtyYWvHAoDtY9NPU3f3nao6meRSkqUk57r7alWdWD9/pqpekuRKkp9M8r2qemeSQ939za0bHQAWw6YxTpLuvpjk4tTamYnH/5G7t68BgBfIN3ABwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGAzxbiqjlTVtaparapTG5yvqvrT9fNPV9Vr5j8qACymTWNcVUtJTid5LMmhJI9X1aGpbY8lObj+czzJn815TgBYWLNcGR9Ostrd17v7dpLzSY5O7Tma5C/6rs8keVFVvXTOswLAQtoxw57dSW5MHK8lee0Me3Yn+fLkpqo6nrtXzknybFVde0HT8oNkV5Kvjh5iq9QToyeAe/Jv74fbyzZanCXGtcFa38eedPfZJGdneE1+wFXVle5eGT0HbDf+7S2mWW5TryXZO3G8J8nN+9gDAGxglhhfTnKwqg5U1c4kx5JcmNpzIcnb1z9V/YtJvtHdX55+IgDguTa9Td3dd6rqZJJLSZaSnOvuq1V1Yv38mSQXk7wxyWqS/0ryO1s3Mj8gvN0AY/i3t4Cq+zlv7QIAD5Fv4AKAwcQYAAYTYwAYbJbfM2abq6pHcvdb1nbn7u+P30xyobs/N3QwgAXhypjnVVXvyt2vQK0k/5i7v+pWSf5yoz8aAjwcVeW3VhaIT1PzvKrq80le2d3fmVrfmeRqdx8cMxlsb1X1pe7eN3oO5sNtajbzvSQ/m+SLU+svXT8HbJGqevpep5K8+GHOwtYSYzbzziSfqKp/zf/9MZB9SV6e5OSooWCbeHGSNyT5+tR6Jfn0wx+HrSLGPK/u/nhVvSJ3/5Tm7tz9n8Baksvd/d2hw8Hi+0iSn+jup6ZPVNUnH/o0bBnvGQPAYD5NDQCDiTEADCbGADCYGAPAYGIMAIP9D5PF40hDkEc9AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 576x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# View gold label distribution.\n",
    "print(df[\"Label\"].value_counts(normalize = True))\n",
    "df[\"Label\"].value_counts(normalize = True).plot.bar(figsize = (8, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "22cf9ca5-b0a9-411d-aeee-9a99b72e2bed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    0.620773\n",
      "1    0.379227\n",
      "Name: Label, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeMAAAD1CAYAAABjj50KAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAM3UlEQVR4nO3dX4ideX3H8c+3kwb6h2Jppmrzx4QaWSIo2GlswVJLWZrVQpQKzVoq/SMhhbR4UTBX3njj4k0pxoYgofSmoVBrg0ZzIVgLKk22bJdmbeyQqpnGYrSibCuN0W8vMrbHs5Ods8mZ/PTM6wWB8/yeH898b5I3z3POnFR3BwAY54dGDwAA250YA8BgYgwAg4kxAAwmxgAwmBgDwGA7Rv3gXbt29f79+0f9eAB46J588smvdPfy9PqwGO/fvz9XrlwZ9eMB4KGrqi9stO4xNQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDDfvSj0W3/9RHRo/AA/j8e944egRgG3FnDACDiTEADDZTjKvqSFVdq6rVqjp1jz2vr6qnqupqVf3dfMcEgMW16XvGVbWU5HSSR5OsJblcVRe6+5mJPS9K8v4kR7r7i1X101s0LwAsnFnujA8nWe3u6919O8n5JEen9rw1yQe7+4tJ0t1fnu+YALC4Zonx7iQ3Jo7X1tcmvSLJT1bVJ6rqyap627wGBIBFN8uvNtUGa73BdX4uya8m+ZEkn66qz3T3577nQlXHkxxPkn379r3waQFgAc1yZ7yWZO/E8Z4kNzfY87Hu/q/u/kqSTyZ59fSFuvtsd69098ry8vL9zgwAC2WWGF9OcrCqDlTVziTHklyY2vO3SX6pqnZU1Y8meW2Sz853VABYTJs+pu7uO1V1MsmlJEtJznX31ao6sX7+THd/tqo+luTpJN9J8oHu/uetHBwAFsVMX4fZ3ReTXJxaOzN1/N4k753faACwPfgGLgAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYbKYYV9WRqrpWVatVdWqD86+vqq9X1VPrf941/1EBYDHt2GxDVS0lOZ3k0SRrSS5X1YXufmZq6993969vwYwAsNBmuTM+nGS1u6939+0k55Mc3dqxAGD7mCXGu5PcmDheW1+b9otV9U9V9dGqeuVcpgOAbWDTx9RJaoO1njr+xyQv6+5nq+oNST6U5OBzLlR1PMnxJNm3b98LmxQAFtQsd8ZrSfZOHO9JcnNyQ3d/o7ufXX99MckPV9Wu6Qt199nuXunuleXl5QcYGwAWxywxvpzkYFUdqKqdSY4luTC5oapeUlW1/vrw+nW/Ou9hAWARbfqYurvvVNXJJJeSLCU5191Xq+rE+vkzSd6S5A+q6k6SbyY51t3Tj7IBgA3M8p7xdx89X5xaOzPx+n1J3jff0QBge/ANXAAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAw2U4yr6khVXauq1ao69Tz7fr6qvl1Vb5nfiACw2DaNcVUtJTmd5LEkh5I8XlWH7rHviSSX5j0kACyyWe6MDydZ7e7r3X07yfkkRzfY94dJ/jrJl+c4HwAsvFlivDvJjYnjtfW1/1NVu5O8OcmZ+Y0GANvDLDGuDdZ66vhPkryzu7/9vBeqOl5VV6rqyq1bt2YcEQAW244Z9qwl2TtxvCfJzak9K0nOV1WS7Eryhqq6090fmtzU3WeTnE2SlZWV6aADwLY0S4wvJzlYVQeS/HuSY0neOrmhuw9893VV/XmSD0+HGADY2KYx7u47VXUydz8lvZTkXHdfraoT6+e9TwwAD2CWO+N098UkF6fWNoxwd//Og48FANuHb+ACgMHEGAAGm+kxNcAPiv2nPjJ6BB7A59/zxtEjDOHOGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhsphhX1ZGqulZVq1V1aoPzR6vq6ap6qqquVNXr5j8qACymHZttqKqlJKeTPJpkLcnlqrrQ3c9MbPt4kgvd3VX1qiR/leSRrRgYABbNLHfGh5Osdvf17r6d5HySo5MbuvvZ7u71wx9L0gEAZjJLjHcnuTFxvLa+9j2q6s1V9S9JPpLk9za6UFUdX3+MfeXWrVv3My8ALJxZYlwbrD3nzre7/6a7H0nypiTv3uhC3X22u1e6e2V5efkFDQoAi2qWGK8l2TtxvCfJzXtt7u5PJvnZqtr1gLMBwLYwS4wvJzlYVQeqameSY0kuTG6oqpdXVa2/fk2SnUm+Ou9hAWARbfpp6u6+U1Unk1xKspTkXHdfraoT6+fPJPmNJG+rqm8l+WaS35z4QBcA8Dw2jXGSdPfFJBen1s5MvH4iyRPzHQ0AtgffwAUAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgM8W4qo5U1bWqWq2qUxuc/62qenr9z6eq6tXzHxUAFtOmMa6qpSSnkzyW5FCSx6vq0NS2f0vyy939qiTvTnJ23oMCwKKa5c74cJLV7r7e3beTnE9ydHJDd3+qu7+2fviZJHvmOyYALK5ZYrw7yY2J47X1tXv5/SQffZChAGA72THDntpgrTfcWPUruRvj193j/PEkx5Nk3759M44IAIttljvjtSR7J473JLk5vamqXpXkA0mOdvdXN7pQd5/t7pXuXlleXr6feQFg4cwS48tJDlbVgarameRYkguTG6pqX5IPJvnt7v7c/McEgMW16WPq7r5TVSeTXEqylORcd1+tqhPr588keVeSn0ry/qpKkjvdvbJ1YwPA4pjlPeN098UkF6fWzky8fnuSt893NADYHnwDFwAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIPNFOOqOlJV16pqtapObXD+kar6dFX9T1X98fzHBIDFtWOzDVW1lOR0kkeTrCW5XFUXuvuZiW3/meSPkrxpK4YEgEU2y53x4SSr3X29u28nOZ/k6OSG7v5yd19O8q0tmBEAFtosMd6d5MbE8dr6GgAwB7PEuDZY6/v5YVV1vKquVNWVW7du3c8lAGDhzBLjtSR7J473JLl5Pz+su89290p3rywvL9/PJQBg4cwS48tJDlbVgarameRYkgtbOxYAbB+bfpq6u+9U1ckkl5IsJTnX3Ver6sT6+TNV9ZIkV5L8RJLvVNU7khzq7m9s3egAsBg2jXGSdPfFJBen1s5MvP6P3H18DQC8QL6BCwAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGmynGVXWkqq5V1WpVndrgfFXVn66ff7qqXjP/UQFgMW0a46paSnI6yWNJDiV5vKoOTW17LMnB9T/Hk/zZnOcEgIU1y53x4SSr3X29u28nOZ/k6NSeo0n+ou/6TJIXVdVL5zwrACykHTPs2Z3kxsTxWpLXzrBnd5IvTW6qquO5e+ecJM9W1bUXNC3fT3Yl+croIbZKPTF6Argnf/d+sL1so8VZYlwbrPV97El3n01ydoafyfe5qrrS3Suj54Dtxt+9xTTLY+q1JHsnjvckuXkfewCADcwS48tJDlbVgarameRYkgtTey4kedv6p6p/IcnXu/tL0xcCAJ5r08fU3X2nqk4muZRkKcm57r5aVSfWz59JcjHJG5KsJvnvJL+7dSPzfcLbDTCGv3sLqLqf89YuAPAQ+QYuABhMjAFgMDEGgMFm+T1jtrmqeiR3v2Vtd+7+/vjNJBe6+7NDBwNYEO6MeV5V9c7c/QrUSvIPufurbpXkLzf6T0OAh6Oq/NbKAvFpap5XVX0uySu7+1tT6zuTXO3ug2Mmg+2tqr7Y3ftGz8F8eEzNZr6T5GeSfGFq/aXr54AtUlVP3+tUkhc/zFnYWmLMZt6R5ONV9a/5//8MZF+Slyc5OWoo2CZenOTXknxtar2SfOrhj8NWEWOeV3d/rKpekbv/lebu3P1HYC3J5e7+9tDhYPF9OMmPd/dT0yeq6hMPfRq2jPeMAWAwn6YGgMHEGAAGE2MAGEyMAWAwMQaAwf4X3bPmCYBVL4kAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 576x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(df_train[\"Label\"].value_counts(normalize = True))\n",
    "df_train[\"Label\"].value_counts(normalize = True).plot.bar(figsize = (8, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "278c0ab3-100d-4bed-8e03-d5b6494ab2be",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    0.621739\n",
      "1    0.378261\n",
      "Name: Label, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeMAAAD1CAYAAABjj50KAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAM3ElEQVR4nO3dX4ideX3H8c+3kwb6h2Jppmrzx4QaWSKsYKexBUstZWlWC1EqNGup9I+EFNLiRcFceeONizelGBuChNKbhkKtDRrNhWAtqDTZsl2atbFDqmYay0YryrbSGP32ImN7PDvZOZs9k5+eeb0gcJ7f8+M535vJm+ecM2equwMAjPNDowcAgO1OjAFgMDEGgMHEGAAGE2MAGEyMAWCwHaOeeNeuXb1///5RTw8AD9wTTzzxle5enl4fFuP9+/fnypUro54eAB64qvriRutepgaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgsGFf+rHo9p/66OgReBG+8N43jR4B2EbcGQPAYGIMAIOJMQAMNlOMq+pIVV2rqtWqOnWPPW+oqier6mpV/d18xwSAxbXpB7iqainJ6SSPJFlLcrmqLnT30xN7XpLkA0mOdPeXquqnt2heAFg4s9wZH06y2t3Xu/t2kvNJjk7teVuSD3X3l5Kku5+Z75gAsLhmifHuJDcmjtfW1ya9KslPVtUnq+qJqnr7vAYEgEU3y+8Z1wZrvcF1fi7Jryb5kSSfqarPdvfnv+dCVceTHE+Sffv2vfBpAWABzXJnvJZk78TxniQ3N9jz8e7+r+7+SpJPJXnN9IW6+2x3r3T3yvLy8v3ODAALZZYYX05ysKoOVNXOJMeSXJja87dJfqmqdlTVjyZ5XZLPzXdUAFhMm75M3d13qupkkktJlpKc6+6rVXVi/fyZ7v5cVX08yVNJvpPkg939z1s5OAAsipm+m7q7Lya5OLV2Zur4fUneN7/RAGB78A1cADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADDZTjKvqSFVdq6rVqjq1wfk3VNXXq+rJ9X/vnv+oALCYdmy2oaqWkpxO8kiStSSXq+pCdz89tfXvu/vXt2BGAFhos9wZH06y2t3Xu/t2kvNJjm7tWACwfcwS491Jbkwcr62vTfvFqvqnqvpYVb16LtMBwDaw6cvUSWqDtZ46/sckr+juZ6vqjUk+nOTgcy5UdTzJ8STZt2/fC5sUABbULHfGa0n2ThzvSXJzckN3f6O7n11/fDHJD1fVrukLdffZ7l7p7pXl5eUXMTYALI5ZYnw5ycGqOlBVO5McS3JhckNVvayqav3x4fXrfnXewwLAItr0ZeruvlNVJ5NcSrKU5Fx3X62qE+vnzyR5a5I/qKo7Sb6Z5Fh3T7+UDQBsYJb3jL/70vPFqbUzE4/fn+T98x0NALYH38AFAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYDPFuKqOVNW1qlqtqlPPs+/nq+rbVfXW+Y0IAItt0xhX1VKS00keTXIoyWNVdege+x5PcmneQwLAIpvlzvhwktXuvt7dt5OcT3J0g31/mOSvkzwzx/kAYOHNEuPdSW5MHK+tr/2fqtqd5C1JzsxvNADYHmaJcW2w1lPHf5LkXd397ee9UNXxqrpSVVdu3bo144gAsNh2zLBnLcneieM9SW5O7VlJcr6qkmRXkjdW1Z3u/vDkpu4+m+RskqysrEwHHQC2pVlifDnJwao6kOTfkxxL8rbJDd194LuPq+rPk3xkOsQAwMY2jXF336mqk7n7KemlJOe6+2pVnVg/731iAHgRZrkzTndfTHJxam3DCHf377z4sQBg+/ANXAAwmBgDwGAzvUwN8INi/6mPjh6BF+EL733T6BGGcGcMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADDZTjKvqSFVdq6rVqjq1wfmjVfVUVT1ZVVeq6vXzHxUAFtOOzTZU1VKS00keSbKW5HJVXejupye2fSLJhe7uqno4yV8leWgrBgaARTPLnfHhJKvdfb27byc5n+To5Ibufra7e/3wx5J0AICZzBLj3UluTByvra99j6p6S1X9S5KPJvm9+YwHAItvlhjXBmvPufPt7r/p7oeSvDnJeza8UNXx9feUr9y6desFDQoAi2qWGK8l2TtxvCfJzXtt7u5PJfnZqtq1wbmz3b3S3SvLy8sveFgAWESzxPhykoNVdaCqdiY5luTC5IaqemVV1frj1ybZmeSr8x4WABbRpp+m7u47VXUyyaUkS0nOdffVqjqxfv5Mkt9I8vaq+laSbyb5zYkPdAEAz2PTGCdJd19McnFq7czE48eTPD7f0QBge/ANXAAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAw2U4yr6khVXauq1ao6tcH536qqp9b/fbqqXjP/UQFgMW0a46paSnI6yaNJDiV5rKoOTW37tyS/3N0PJ3lPkrPzHhQAFtUsd8aHk6x29/Xuvp3kfJKjkxu6+9Pd/bX1w88m2TPfMQFgcc0S491Jbkwcr62v3cvvJ/nYixkKALaTHTPsqQ3WesONVb+SuzF+/T3OH09yPEn27ds344gAsNhmuTNeS7J34nhPkpvTm6rq4SQfTHK0u7+60YW6+2x3r3T3yvLy8v3MCwALZ5YYX05ysKoOVNXOJMeSXJjcUFX7knwoyW939+fnPyYALK5NX6bu7jtVdTLJpSRLSc5199WqOrF+/kySdyf5qSQfqKokudPdK1s3NgAsjlneM053X0xycWrtzMTjdyR5x3xHA4DtwTdwAcBgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMNhMMa6qI1V1rapWq+rUBucfqqrPVNX/VNUfz39MAFhcOzbbUFVLSU4neSTJWpLLVXWhu5+e2PafSf4oyZu3YkgAWGSz3BkfTrLa3de7+3aS80mOTm7o7me6+3KSb23BjACw0GaJ8e4kNyaO19bXAIA5mCXGtcFa38+TVdXxqrpSVVdu3bp1P5cAgIUzS4zXkuydON6T5Ob9PFl3n+3ule5eWV5evp9LAMDCmSXGl5McrKoDVbUzybEkF7Z2LADYPjb9NHV336mqk0kuJVlKcq67r1bVifXzZ6rqZUmuJPmJJN+pqncmOdTd39i60QFgMWwa4yTp7otJLk6tnZl4/B+5+/I1APAC+QYuABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhsphhX1ZGqulZVq1V1aoPzVVV/un7+qap67fxHBYDFtGmMq2opyekkjyY5lOSxqjo0te3RJAfX/x1P8mdznhMAFtYsd8aHk6x29/Xuvp3kfJKjU3uOJvmLvuuzSV5SVS+f86wAsJB2zLBnd5IbE8drSV43w57dSb48uamqjufunXOSPFtV117QtHw/2ZXkK6OH2Cr1+OgJ4J787P1ge8VGi7PEuDZY6/vYk+4+m+TsDM/J97mqutLdK6PngO3Gz95imuVl6rUkeyeO9yS5eR97AIANzBLjy0kOVtWBqtqZ5FiSC1N7LiR5+/qnqn8hyde7+8vTFwIAnmvTl6m7+05VnUxyKclSknPdfbWqTqyfP5PkYpI3JllN8t9JfnfrRub7hLcbYAw/ewuoup/z1i4A8AD5Bi4AGEyMAWAwMQaAwWb5PWO2uap6KHe/ZW137v7++M0kF7r7c0MHA1gQ7ox5XlX1rtz9CtRK8g+5+6tuleQvN/qjIcCDUVV+a2WB+DQ1z6uqPp/k1d39ran1nUmudvfBMZPB9lZVX+rufaPnYD68TM1mvpPkZ5J8cWr95evngC1SVU/d61SSlz7IWdhaYsxm3pnkE1X1r/n/PwayL8krk5wcNRRsEy9N8mtJvja1Xkk+/eDHYauIMc+ruz9eVa/K3T+luTt3/xNYS3K5u789dDhYfB9J8uPd/eT0iar65AOfhi3jPWMAGMynqQFgMDEGgMHEGAAGE2MAGEyMAWCw/wUQUeYJhF2cagAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 576x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(df_test[\"Label\"].value_counts(normalize = True))\n",
    "df_test[\"Label\"].value_counts(normalize = True).plot.bar(figsize = (8, 4))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "106ebdc9-5e32-482f-a04a-bc4d08612ef7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0    0.619565\n",
      "1    0.380435\n",
      "Name: Label, dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:>"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAeMAAAD1CAYAAABjj50KAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAM10lEQVR4nO3dX4ice33H8c+3SQP9Q7E0W7X5Y0KNHCIo2G1swVJLOTRHC1EqNFoq/SMhhbR4UTBX3njjwZvSGhuChNKbhkKtDRrNhWAtqJiccnpojo1dUjXbWIxWFFtpjH57ka0dx8nZOTmz+ens6wWBeX7Pj2e+N5s3z8zsbHV3AIBxfmj0AACw3YkxAAwmxgAwmBgDwGBiDACDiTEADLZz1BPv3r27Dxw4MOrpAeChe+KJJ77U3SvT68NifODAgVy9enXU0wPAQ1dVn5u17mVqABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYb9qUfy+7A6Q+OHoHn4LPvfO3oEYBtxJ0xAAwmxgAw2FwxrqqjVXW9qtaq6vR99ry6qp6sqmtV9feLHRMAltem7xlX1Y4kZ5I8mmQ9yZWqutjdT0/seV6S9yQ52t2fr6qf3qJ5AWDpzHNnfCTJWnff6O47SS4kOTa1501J3tfdn0+S7v7iYscEgOU1T4z3JLk5cby+sTbpJUl+sqo+WlVPVNWbFzUgACy7eX61qWas9Yzr/FySX03yI0k+UVWf7O7PfNeFqk4kOZEk+/fvf/bTAsASmufOeD3JvonjvUluzdjz4e7+r+7+UpKPJXn59IW6+1x3r3b36srKyoPODABLZZ4YX0lyqKoOVtWuJMeTXJza83dJfqmqdlbVjyZ5ZZJPL3ZUAFhOm75M3d13q+pUkstJdiQ5393Xqurkxvmz3f3pqvpwkqeSfDvJe7v7n7dycABYFnN9HWZ3X0pyaWrt7NTxu5K8a3GjAcD24Bu4AGAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWCwuWJcVUer6npVrVXV6RnnX11VX62qJzf+vX3xowLActq52Yaq2pHkTJJHk6wnuVJVF7v76amt/9Ddv74FMwLAUpvnzvhIkrXuvtHdd5JcSHJsa8cCgO1jnhjvSXJz4nh9Y23aL1bVP1XVh6rqpbMuVFUnqupqVV29ffv2A4wLAMtnnhjXjLWeOv7HJC/q7pcn+bMk7591oe4+192r3b26srLyrAYFgGU1T4zXk+ybON6b5Nbkhu7+Wnd/fePxpSQ/XFW7FzYlACyxeWJ8JcmhqjpYVbuSHE9ycXJDVb2gqmrj8ZGN63550cMCwDLa9NPU3X23qk4luZxkR5Lz3X2tqk5unD+b5A1J/qCq7ib5RpLj3T39UjYAMMOmMU6+89Lzpam1sxOP353k3YsdDQC2B9/ABQCDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBzxbiqjlbV9apaq6rTz7Dv56vqW1X1hsWNCADLbdMYV9WOJGeSPJbkcJI3VtXh++x7PMnlRQ8JAMtsnjvjI0nWuvtGd99JciHJsRn7/jDJ3yT54gLnA4ClN0+M9yS5OXG8vrH2HVW1J8nrk5xd3GgAsD3ME+OasdZTx3+S5G3d/a1nvFDViaq6WlVXb9++PeeIALDcds6xZz3JvonjvUluTe1ZTXKhqpJkd5LXVNXd7n7/5KbuPpfkXJKsrq5OBx0AtqV5YnwlyaGqOpjk35McT/KmyQ3dffD/HlfVXyT5wHSIAYDZNo1xd9+tqlO59ynpHUnOd/e1qjq5cd77xADwHMxzZ5zuvpTk0tTazAh39+8897EAYPvwDVwAMNhcd8YAPygOnP7g6BF4Dj77zteOHmEId8YAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYHPFuKqOVtX1qlqrqtMzzh+rqqeq6smqulpVr1r8qACwnHZutqGqdiQ5k+TRJOtJrlTVxe5+emLbR5Jc7O6uqpcl+eskj2zFwACwbOa5Mz6SZK27b3T3nSQXkhyb3NDdX+/u3jj8sSQdAGAu88R4T5KbE8frG2vfpapeX1X/kuSDSX5vMeMBwPKbJ8Y1Y+177ny7+2+7+5Ekr0vyjpkXqjqx8Z7y1du3bz+rQQFgWc0T4/Uk+yaO9ya5db/N3f2xJD9bVbtnnDvX3avdvbqysvKshwWAZTRPjK8kOVRVB6tqV5LjSS5ObqiqF1dVbTx+RZJdSb686GEBYBlt+mnq7r5bVaeSXE6yI8n57r5WVSc3zp9N8htJ3lxV30zyjSS/OfGBLgDgGWwa4yTp7ktJLk2tnZ14/HiSxxc7GgBsD76BCwAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGE2MAGEyMAWAwMQaAwcQYAAYTYwAYTIwBYDAxBoDBxBgABhNjABhMjAFgMDEGgMHEGAAGmyvGVXW0qq5X1VpVnZ5x/req6qmNfx+vqpcvflQAWE6bxriqdiQ5k+SxJIeTvLGqDk9t+7ckv9zdL0vyjiTnFj0oACyree6MjyRZ6+4b3X0nyYUkxyY3dPfHu/srG4efTLJ3sWMCwPKaJ8Z7ktycOF7fWLuf30/yoecyFABsJzvn2FMz1nrmxqpfyb0Yv+o+508kOZEk+/fvn3NEAFhu89wZryfZN3G8N8mt6U1V9bIk701yrLu/POtC3X2uu1e7e3VlZeVB5gWApTNPjK8kOVRVB6tqV5LjSS5Obqiq/Unel+S3u/szix8TAJbXpi9Td/fdqjqV5HKSHUnOd/e1qjq5cf5skrcn+akk76mqJLnb3atbNzYALI953jNOd19Kcmlq7ezE47ckectiRwOA7cE3cAHAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADDYXDGuqqNVdb2q1qrq9Izzj1TVJ6rqf6rqjxc/JgAsr52bbaiqHUnOJHk0yXqSK1V1sbufntj2n0n+KMnrtmJIAFhm89wZH0my1t03uvtOkgtJjk1u6O4vdveVJN/cghkBYKnNE+M9SW5OHK9vrAEACzBPjGvGWj/Ik1XViaq6WlVXb9++/SCXAIClM0+M15Psmzjem+TWgzxZd5/r7tXuXl1ZWXmQSwDA0pknxleSHKqqg1W1K8nxJBe3diwA2D42/TR1d9+tqlNJLifZkeR8d1+rqpMb589W1QuSXE3yE0m+XVVvTXK4u7+2daMDwHLYNMZJ0t2XklyaWjs78fg/cu/lawDgWfINXAAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAwmBgDwGBiDACDiTEADCbGADCYGAPAYGIMAIOJMQAMJsYAMJgYA8BgYgwAg4kxAAwmxgAw2FwxrqqjVXW9qtaq6vSM81VVf7px/qmqesXiRwWA5bRpjKtqR5IzSR5LcjjJG6vq8NS2x5Ic2vh3IsmfL3hOAFha89wZH0my1t03uvtOkgtJjk3tOZbkL/ueTyZ5XlW9cMGzAsBS2jnHnj1Jbk4cryd55Rx79iT5wuSmqjqRe3fOSfL1qrr+rKbl+8nuJF8aPcRWqcdHTwD35WfvB9uLZi3OE+OasdYPsCfdfS7JuTmek+9zVXW1u1dHzwHbjZ+95TTPy9TrSfZNHO9NcusB9gAAM8wT4ytJDlXVwaraleR4kotTey4mefPGp6p/IclXu/sL0xcCAL7Xpi9Td/fdqjqV5HKSHUnOd/e1qjq5cf5skktJXpNkLcl/J/ndrRuZ7xPeboAx/Owtoer+nrd2AYCHyDdwAcBgYgwAg4kxAAw2z+8Zs81V1SO59y1re3Lv98dvJbnY3Z8eOhjAknBnzDOqqrfl3legVpJP5d6vulWSv5r1R0OAh6Oq/NbKEvFpap5RVX0myUu7+5tT67uSXOvuQ2Mmg+2tqj7f3ftHz8FieJmazXw7yc8k+dzU+gs3zgFbpKqeut+pJM9/mLOwtcSYzbw1yUeq6l/z/38MZH+SFyc5NWoo2Caen+TXknxlar2SfPzhj8NWEWOeUXd/uKpeknt/SnNP7v0nsJ7kSnd/a+hwsPw+kOTHu/vJ6RNV9dGHPg1bxnvGADCYT1MDwGBiDACDiTEADCbGADCYGAPAYP8LDcrpAriODNUAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 576x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(df_val[\"Label\"].value_counts(normalize = True))\n",
    "df_val[\"Label\"].value_counts(normalize = True).plot.bar(figsize = (8, 4))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91caec48-6839-4e1f-b32a-4bcf63765253",
   "metadata": {},
   "source": [
    "## Labeling functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c681538f-de23-409a-a9fd-311493e99fed",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Label macros.\n",
    "ABSTAIN = -1\n",
    "EXCLUDE = 0\n",
    "INCLUDE = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "09fbb065-a9fd-4055-bfff-31f5894f43fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "89\n",
      "['telomer', 'hairpin', 'salmon', 'snake', 'rna replication', 'pig', 'population dynamics', 'composition dynamics', 'rna world', 'chicken', 'livestock', 'reptile', 'avian', 'reporter', 'frog', 'rna structur', 'cattle', 'viral dynamics', 'heifer', 'epilep', 'finch', 'crab', 'regulat', 'conformation', 'transcription factor', 'ovine', 'md', 'ribosom', 'ciona', 'fluoresc', 'mallard', 'spin', 'chromatin dynamics', 'switch', 'rna metaboli', 'goat', 'chordate', 'insect', 'nmr', 'duck', 'amphibian', 'rna decay', 'cytometry', 'rna degradation', 'molecular dynamic', 'cow', 'sheep', 'bos', 'temporal lobe', 'wastewater', 'fish', 'spectr', 'noncoding', 'folding', 'cichlid', 'rna quality', 'kinetic', 'imaging', 'porcine', 'aquatic', 'arachnid', 'telomerase rna', 'buckl', 'energy', 'xenopus', 'bird', 'replication system', 'caprine', 'promoter', 'structural dynamics', 'trout', 'dynamic modularity', 'crustacean', 'spectroscop', 'protein', 'water', 'marine', 'evolution', 'repuckering dynamics', 'imag', 'non-coding', 'developmental dynamics', 'secondary structur', 'polymerase', 'viral load', 'control', 'bovine', 'yak', 'loop']\n"
     ]
    }
   ],
   "source": [
    "# Data for labeling functions.\n",
    "unwanted_animals = [\"xenopus\", \"marine\", \"aquatic\", \"cow\", \"Bos\", \"bovine\", \"cattle\", \"sheep\", \n",
    "                   \"ovine\", \"caprine\", \"goat\", \"livestock\", \"pig\", \"porcine\", \"amphibian\", \n",
    "                   \"reptile\", \"frog\", \"snake\", \"fish\", \"cichlid\", \"crustacean\", \"crab\", \"insect\", \n",
    "                   \"chicken\", \"duck\", \"mallard\", \"finch\", \"bird\", \"avian\", \"arachnid\", \"yak\",\n",
    "                   \"chordate\", \"Ciona\", \"trout\", \"salmon\", \"heifer\"]\n",
    "unwanted_terms = [\"NMR\", \"MD\", \"molecular dynamic\", \"RNA world\", \"wastewater\", \n",
    "                 \"water\", \"viral load\", \"RNA degradation\", \"RNA structur\", \"secondary structur\", \n",
    "                 \"folding\", \"regulat\", \"promoter\", \"spectroscop\", \"spectr\", \"RNA decay\", \"RNA replication\", \n",
    "                 \"transcription factor\", \"switch\", \"control\", \"RNA metaboli\", \n",
    "                 \"population dynamics\", \"evolution\", \"protein\", \"conformation\", \"viral dynamics\", \n",
    "                 \"polymerase\", \"telomerase RNA\", \"Structural dynamics\", \"control\", \"ribosom\", \n",
    "                 \"repuckering dynamics\", \"replication system\", \"loop\", \"hairpin\", \"chromatin dynamics\", \n",
    "                 \"dynamic modularity\", \"imaging\", \"reporter\", \"kinetic\", \n",
    "                 \"developmental dynamics\", \"temporal lobe\", \"epilep\", \n",
    "                 \"cytometry\", \"imag\", \"fluoresc\", \"telomer\", \"composition dynamics\", \"noncoding\", \n",
    "                 \"non-coding\", \"buckl\", \"spin\", \"energy\", \"RNA quality\"]\n",
    "necessary_terms = [\"rna-seq\", \"microarray\", \"gene expression\", \"transcriptom\"]\n",
    "optional_terms = [\"longitudinal\", \"timecourse\", \"time-course\", \"time series\"]\n",
    "unwanted = [x.lower() for x in unwanted_animals + unwanted_terms]\n",
    "unwanted = set(unwanted)\n",
    "unwanted = list(unwanted)\n",
    "print(len(unwanted))\n",
    "print(unwanted)\n",
    "#any(term in \"ribosomal\" for term in unwanted_terms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "6c56a70e-8478-4687-8042-bdac000f90a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "@labeling_function()\n",
    "def exclude_animals(x):\n",
    "    return EXCLUDE if any(animal in x.Title for animal in unwanted_animals) else ABSTAIN\n",
    "\n",
    "@labeling_function()\n",
    "def exclude_terms(x):\n",
    "    return EXCLUDE if any(term in x.Title for term in unwanted_terms) else ABSTAIN\n",
    "\n",
    "@labeling_function()\n",
    "def include_necessary_terms(x):\n",
    "    return INCLUDE if any(term in x.Title for term in necessary_terms) else EXCLUDE"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7e186e6-5b55-42d8-a953-73acdb04e2bd",
   "metadata": {},
   "source": [
    "## Apply labeling functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "0e06f424-1595-415c-ac59-e062efbf3789",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████| 1656/1656 [00:00<00:00, 2270.34it/s]\n",
      "100%|███████████████████████████████████████| 184/184 [00:00<00:00, 2238.06it/s]\n",
      "100%|███████████████████████████████████████| 460/460 [00:00<00:00, 2176.29it/s]\n"
     ]
    }
   ],
   "source": [
    "# Apply LFs to dataframe.\n",
    "lfs = [exclude_animals,\n",
    "       exclude_terms,\n",
    "       include_necessary_terms]\n",
    "applier = PandasLFApplier(lfs = lfs)\n",
    "L_train = applier.apply(df = X_train)\n",
    "L_val = applier.apply(df = X_val)\n",
    "L_test = applier.apply(df = X_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ac65f849-308f-4444-a89a-c0a3324c7510",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "      <th>Correct</th>\n",
       "      <th>Incorrect</th>\n",
       "      <th>Emp. Acc.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exclude_animals</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.052536</td>\n",
       "      <td>0.052536</td>\n",
       "      <td>0.043478</td>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exclude_terms</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.374396</td>\n",
       "      <td>0.374396</td>\n",
       "      <td>0.173309</td>\n",
       "      <td>620</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>include_necessary_terms</th>\n",
       "      <td>2</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.413647</td>\n",
       "      <td>0.207126</td>\n",
       "      <td>1127</td>\n",
       "      <td>529</td>\n",
       "      <td>0.680556</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         j Polarity  Coverage  Overlaps  Conflicts  Correct  \\\n",
       "exclude_animals          0      [0]  0.052536  0.052536   0.043478       87   \n",
       "exclude_terms            1      [0]  0.374396  0.374396   0.173309      620   \n",
       "include_necessary_terms  2   [0, 1]  1.000000  0.413647   0.207126     1127   \n",
       "\n",
       "                         Incorrect  Emp. Acc.  \n",
       "exclude_animals                  0   1.000000  \n",
       "exclude_terms                    0   1.000000  \n",
       "include_necessary_terms        529   0.680556  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TRAIN: Summarize coverage, conflicts, empirical accurcacy of LFs.\n",
    "LFAnalysis(L_train, lfs).lf_summary(y_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "da4c0fb9-f136-4f4d-80f8-e2c84f6b34b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "      <th>Correct</th>\n",
       "      <th>Incorrect</th>\n",
       "      <th>Emp. Acc.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exclude_animals</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.048913</td>\n",
       "      <td>0.048913</td>\n",
       "      <td>0.027174</td>\n",
       "      <td>9</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exclude_terms</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.423913</td>\n",
       "      <td>0.423913</td>\n",
       "      <td>0.222826</td>\n",
       "      <td>78</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>include_necessary_terms</th>\n",
       "      <td>2</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.451087</td>\n",
       "      <td>0.239130</td>\n",
       "      <td>121</td>\n",
       "      <td>63</td>\n",
       "      <td>0.657609</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         j Polarity  Coverage  Overlaps  Conflicts  Correct  \\\n",
       "exclude_animals          0      [0]  0.048913  0.048913   0.027174        9   \n",
       "exclude_terms            1      [0]  0.423913  0.423913   0.222826       78   \n",
       "include_necessary_terms  2   [0, 1]  1.000000  0.451087   0.239130      121   \n",
       "\n",
       "                         Incorrect  Emp. Acc.  \n",
       "exclude_animals                  0   1.000000  \n",
       "exclude_terms                    0   1.000000  \n",
       "include_necessary_terms         63   0.657609  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# VAL: Summarize coverage, conflicts, empirical accurcacy of LFs.\n",
    "LFAnalysis(L_val, lfs).lf_summary(y_val)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "65ba588d-1ae3-4208-a3e6-28aefdf2e54f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n",
      "/opt/anaconda3/lib/python3.8/site-packages/sklearn/utils/validation.py:70: FutureWarning: Pass labels=[-1  0  1] as keyword args. From version 1.0 (renaming of 0.25) passing these as positional arguments will result in an error\n",
      "  warnings.warn(f\"Pass {args_msg} as keyword args. From version \"\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>j</th>\n",
       "      <th>Polarity</th>\n",
       "      <th>Coverage</th>\n",
       "      <th>Overlaps</th>\n",
       "      <th>Conflicts</th>\n",
       "      <th>Correct</th>\n",
       "      <th>Incorrect</th>\n",
       "      <th>Emp. Acc.</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>exclude_animals</th>\n",
       "      <td>0</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.063043</td>\n",
       "      <td>0.063043</td>\n",
       "      <td>0.058696</td>\n",
       "      <td>29</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exclude_terms</th>\n",
       "      <td>1</td>\n",
       "      <td>[0]</td>\n",
       "      <td>0.358696</td>\n",
       "      <td>0.358696</td>\n",
       "      <td>0.191304</td>\n",
       "      <td>165</td>\n",
       "      <td>0</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>include_necessary_terms</th>\n",
       "      <td>2</td>\n",
       "      <td>[0, 1]</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.406522</td>\n",
       "      <td>0.239130</td>\n",
       "      <td>302</td>\n",
       "      <td>158</td>\n",
       "      <td>0.656522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                         j Polarity  Coverage  Overlaps  Conflicts  Correct  \\\n",
       "exclude_animals          0      [0]  0.063043  0.063043   0.058696       29   \n",
       "exclude_terms            1      [0]  0.358696  0.358696   0.191304      165   \n",
       "include_necessary_terms  2   [0, 1]  1.000000  0.406522   0.239130      302   \n",
       "\n",
       "                         Incorrect  Emp. Acc.  \n",
       "exclude_animals                  0   1.000000  \n",
       "exclude_terms                    0   1.000000  \n",
       "include_necessary_terms        158   0.656522  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# TEST: Summarize coverage, conflicts, empirical accurcacy of LFs.\n",
    "LFAnalysis(L_test, lfs).lf_summary(y_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "69a85077-ebe4-4c04-8dd4-dcf5e45764fa",
   "metadata": {},
   "source": [
    "## Dummy accuracy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b2594977-7d48-49bb-8844-a22f7f8c6926",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Accuracy if predicting majority class 0.6217391304347826\n"
     ]
    }
   ],
   "source": [
    "# View \"dummy\" accuracy if predicting majority class every time.\n",
    "print(\"Accuracy if predicting majority class\", \n",
    "      df_test[\"Label\"].value_counts(normalize = True).max())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d7058509-2b3c-4f07-8e08-18270ac595aa",
   "metadata": {},
   "source": [
    "## Majority vote"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "61321a43-cb4c-4c5c-b795-e602a27c06bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Majority vote model.\n",
    "majority_model = MajorityLabelVoter()\n",
    "preds_train = majority_model.predict(L = L_train)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6c75149b-58c1-462f-8523-0d790cb5c8fe",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n"
     ]
    }
   ],
   "source": [
    "# Compute majority vote performance metrics.\n",
    "majority_scores = majority_model.score(L = L_test, \n",
    "                                       Y = y_test, \n",
    "                                       tie_break_policy = \"abstain\",\n",
    "                                       metrics = [\"f1\", \"accuracy\", \"precision\", \n",
    "                                                  \"recall\", \"roc_auc\", \"coverage\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "c3324bfc-fbb3-40f5-af5c-7833e368fed7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Majority Model F1:        87.3%\n",
      "Majority Model Accuracy:  86.5%\n",
      "Majority Model Precision: 80.9%\n",
      "Majority Model Recall:    94.8%\n",
      "Majority Model AUC ROC:   86.6%\n",
      "Majority Model Coverage:  77.2%\n"
     ]
    }
   ],
   "source": [
    "# Compare model performance metrics.\n",
    "majority_f1 = majority_scores.get(\"f1\")\n",
    "majority_acc = majority_scores.get(\"accuracy\")\n",
    "majority_prec = majority_scores.get(\"precision\")\n",
    "majority_rec = majority_scores.get(\"recall\")\n",
    "majority_roc = majority_scores.get(\"roc_auc\")\n",
    "majority_cov = majority_scores.get(\"coverage\")\n",
    "print(f\"{'Majority Model F1:':<25} {majority_f1 * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Accuracy:':<25} {majority_acc * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Precision:':<25} {majority_prec * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Recall:':<25} {majority_rec * 100:.1f}%\")\n",
    "print(f\"{'Majority Model AUC ROC:':<25} {majority_roc * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Coverage:':<25} {majority_cov * 100:.1f}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23c7cc77-f1a1-4e17-951f-79a58daf387e",
   "metadata": {},
   "source": [
    "## Model tuning\n",
    "\n",
    "Tune labeling model [hyperparameters](https://snorkel.readthedocs.io/en/v0.9.7/packages/_autosummary/labeling/snorkel.labeling.model.label_model.LabelModel.html#snorkel.labeling.model.label_model.LabelModel) using grid search."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "519eccc7-b2c6-4d34-8dd4-e7adec24e5c7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETER SEARCH SPACE: ---\n",
      "\n",
      "Total combinations: 27\n",
      "\n",
      "[(50, 0.0, 0.001), (50, 0.0, 0.01), (50, 0.0, 0.1), (50, 0.2, 0.001), (50, 0.2, 0.01), (50, 0.2, 0.1), (50, 0.4, 0.001), (50, 0.4, 0.01), (50, 0.4, 0.1), (100, 0.0, 0.001), (100, 0.0, 0.01), (100, 0.0, 0.1), (100, 0.2, 0.001), (100, 0.2, 0.01), (100, 0.2, 0.1), (100, 0.4, 0.001), (100, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.0, 0.001), (250, 0.0, 0.01), (250, 0.0, 0.1), (250, 0.2, 0.001), (250, 0.2, 0.01), (250, 0.2, 0.1), (250, 0.4, 0.001), (250, 0.4, 0.01), (250, 0.4, 0.1)]\n"
     ]
    }
   ],
   "source": [
    "# Search space.\n",
    "epochs = [50, 100, 250]\n",
    "l2 = [0.0, 0.2, 0.4]\n",
    "lr = [0.001, 0.01, 0.1]\n",
    "\n",
    "# Take cartesian product to obtain grid search space.\n",
    "search_space = list(itertools.product(epochs, l2, lr))\n",
    "print(\"\\n--- HYPERPARAMETER SEARCH SPACE: ---\\n\")\n",
    "print(\"Total combinations:\", len(search_space))\n",
    "print()\n",
    "print(search_space)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "29029239-cbc5-460f-a5ce-476ac782e55f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 50/50 [00:00<00:00, 149.39epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|██████████████████████████████████████| 50/50 [00:00<00:00, 1080.94epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|███████████████████████████████████████| 50/50 [00:00<00:00, 992.95epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|███████████████████████████████████████| 50/50 [00:00<00:00, 965.53epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.0, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.0, 0.01)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.0, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 50/50 [00:00<00:00, 997.09epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|███████████████████████████████████████| 50/50 [00:00<00:00, 947.25epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|██████████████████████████████████████| 50/50 [00:00<00:00, 1041.88epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.2, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.2, 0.01)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.2, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████████████████████████████████| 50/50 [00:00<00:00, 1010.89epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|██████████████████████████████████████| 50/50 [00:00<00:00, 1068.14epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/100 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.4, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.4, 0.01)\n",
      "F1: 85.9 | Accuracy: 88.04 | Precision: 77.91 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(50, 0.4, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1109.26epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1046.35epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/100 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.0, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.0, 0.01)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1047.77epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1026.89epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/100 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.0, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.2, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1002.17epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1096.48epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/100 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.2, 0.01)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.2, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1013.37epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|█████████████████████████████████████| 100/100 [00:00<00:00, 984.13epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/100 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.4, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.4, 0.01)\n",
      "F1: 85.9 | Accuracy: 88.04 | Precision: 77.91 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 100/100 [00:00<00:00, 1001.60epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      " 41%|██████████████▊                     | 103/250 [00:00<00:00, 1028.62epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(100, 0.4, 0.1)\n",
      "F1: 85.9 | Accuracy: 88.04 | Precision: 77.91 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1083.71epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      " 49%|█████████████████▋                  | 123/250 [00:00<00:00, 1228.10epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.0, 0.001)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1253.72epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      " 44%|███████████████▊                    | 110/250 [00:00<00:00, 1095.07epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.0, 0.01)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|█████████████████████████████████████| 250/250 [00:00<00:00, 984.72epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1455.01epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/250 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.0, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.2, 0.001)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1390.64epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1467.44epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      "  0%|                                                | 0/250 [00:00<?, ?epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.2, 0.01)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n",
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.2, 0.1)\n",
      "F1: 84.81 | Accuracy: 86.96 | Precision: 76.14 | Recall: 95.71 | AUC ROC: 88.15 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1251.46epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      " 45%|████████████████▎                   | 113/250 [00:00<00:00, 1115.67epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.4, 0.001)\n",
      "F1: 68.72 | Accuracy: 66.85 | Precision: 53.6 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1167.04epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n",
      " 44%|███████████████▉                    | 111/250 [00:00<00:00, 1109.41epoch/s]"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.4, 0.01)\n",
      "F1: 85.9 | Accuracy: 88.04 | Precision: 77.91 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|████████████████████████████████████| 250/250 [00:00<00:00, 1151.19epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS (epochs, l2, lr): ---\n",
      "(250, 0.4, 0.1)\n",
      "F1: 85.9 | Accuracy: 88.04 | Precision: 77.91 | Recall: 95.71 | AUC ROC: 90.75 | Coverage: 100.0\n",
      "----------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "# Validation loop.\n",
    "# Select optimal model based on minimizing false negatives,\n",
    "# i.e. look at recall (syn. sensitivity, TPR), f1, etc.\n",
    "recalls = []\n",
    "f1s = []\n",
    "rocs = []\n",
    "accuracies = []\n",
    "precisions = []\n",
    "for hparams in search_space:\n",
    "    \n",
    "    # Extract hyperparameter values.\n",
    "    n_epochs = hparams[0]\n",
    "    l2 = hparams[1]\n",
    "    lr = hparams[2]\n",
    "    \n",
    "    # Label model.\n",
    "    label_model = LabelModel(cardinality = 2, verbose = True)\n",
    "    label_model.fit(L_train = L_train, \n",
    "                    #Y_dev = y_val,\n",
    "                    #class_balance = [0.7, 0.3], \n",
    "                    n_epochs = n_epochs, \n",
    "                    l2 = l2,\n",
    "                    lr = lr,\n",
    "                    optimizer = \"sgd\",\n",
    "                    seed = seed)\n",
    "\n",
    "    # Compute model performance metrics.\n",
    "    label_scores = label_model.score(L = L_val, \n",
    "                                     Y = y_val, \n",
    "                                     tie_break_policy = \"abstain\",\n",
    "                                     metrics = [\"f1\", \"accuracy\", \"precision\", \n",
    "                                                \"recall\", \"roc_auc\", \"coverage\"])\n",
    "    \n",
    "    \n",
    "\n",
    "    \n",
    "    print(\"\\n--- HYPERPARAMETERS (epochs, l2, lr): ---\")\n",
    "    print(hparams)\n",
    "\n",
    "    label_f1 = label_scores.get(\"f1\") * 100\n",
    "    label_acc = label_scores.get(\"accuracy\") * 100\n",
    "    label_prec = label_scores.get(\"precision\") * 100\n",
    "    label_rec = label_scores.get(\"recall\") * 100\n",
    "    label_roc = label_scores.get(\"roc_auc\") * 100\n",
    "    label_cov = label_scores.get(\"coverage\") * 100\n",
    "    print(\"F1: {} | Accuracy: {} | Precision: {} | Recall: {} | AUC ROC: {} | Coverage: {}\".format(round(label_f1, 2),\n",
    "                                                                                                   round(label_acc, 2),\n",
    "                                                                                                   round(label_prec, 2),\n",
    "                                                                                                   round(label_rec, 2),\n",
    "                                                                                                   round(label_roc, 2),\n",
    "                                                                                                   round(label_cov, 2)))\n",
    "    print(\"----------------------------------------------------\")\n",
    "    \n",
    "    recalls.append(label_rec)\n",
    "    f1s.append(label_f1)\n",
    "    rocs.append(label_roc)\n",
    "    accuracies.append(label_acc)\n",
    "    precisions.append(label_prec)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ba4d0d1d-3173-4aa1-a90b-4409317f01eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "--- HYPERPARAMETERS FOR MAX RECALL OF 95.71428571428572: ---\n",
      "\n",
      "Total models with optimal recall: 27\n",
      "\n",
      "[(50, 0.0, 0.001), (50, 0.0, 0.01), (50, 0.0, 0.1), (50, 0.2, 0.001), (50, 0.2, 0.01), (50, 0.2, 0.1), (50, 0.4, 0.001), (50, 0.4, 0.01), (50, 0.4, 0.1), (100, 0.0, 0.001), (100, 0.0, 0.01), (100, 0.0, 0.1), (100, 0.2, 0.001), (100, 0.2, 0.01), (100, 0.2, 0.1), (100, 0.4, 0.001), (100, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.0, 0.001), (250, 0.0, 0.01), (250, 0.0, 0.1), (250, 0.2, 0.001), (250, 0.2, 0.01), (250, 0.2, 0.1), (250, 0.4, 0.001), (250, 0.4, 0.01), (250, 0.4, 0.1)]\n",
      "\n",
      "--- HYPERPARAMETERS FOR MAX F1 OF 85.8974358974359: ---\n",
      "\n",
      "Total models with optimal F1: 5\n",
      "\n",
      "[(50, 0.4, 0.01), (100, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.4, 0.01), (250, 0.4, 0.1)]\n",
      "\n",
      "--- HYPERPARAMETERS FOR MAX ACCURACY OF 88.04347826086956: ---\n",
      "\n",
      "Total models with optimal accuracy: 5\n",
      "\n",
      "[(50, 0.4, 0.01), (100, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.4, 0.01), (250, 0.4, 0.1)]\n",
      "\n",
      "--- HYPERPARAMETERS FOR MAX PRECISION OF 77.90697674418605: ---\n",
      "\n",
      "Total models with optimal precision: 5\n",
      "\n",
      "[(50, 0.4, 0.01), (100, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.4, 0.01), (250, 0.4, 0.1)]\n",
      "\n",
      "--- HYPERPARAMETERS FOR MAX AUC ROC OF 90.74561403508771: ---\n",
      "\n",
      "Total models with optimal AUC ROC: 12\n",
      "\n",
      "[(50, 0.0, 0.001), (50, 0.2, 0.001), (50, 0.4, 0.001), (50, 0.4, 0.01), (100, 0.0, 0.001), (100, 0.2, 0.001), (100, 0.4, 0.001), (100, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.4, 0.001), (250, 0.4, 0.01), (250, 0.4, 0.1)]\n"
     ]
    }
   ],
   "source": [
    "# Get hyparameter combinations with maximum metrics.\n",
    "max_recall = max(recalls)\n",
    "max_recall_indices = [i for i,j in enumerate(recalls) if j == max_recall]\n",
    "hparams_max_recall = [search_space[i] for i in max_recall_indices]\n",
    "print(\"\\n--- HYPERPARAMETERS FOR MAX RECALL OF {}: ---\\n\".format(max_recall))\n",
    "print(\"Total models with optimal recall:\", len(hparams_max_recall))\n",
    "print()\n",
    "print(hparams_max_recall)\n",
    "\n",
    "max_f1 = max(f1s)\n",
    "max_f1_indices = [i for i,j in enumerate(f1s) if j == max_f1]\n",
    "hparams_max_f1 = [search_space[i] for i in max_f1_indices]\n",
    "print(\"\\n--- HYPERPARAMETERS FOR MAX F1 OF {}: ---\\n\".format(max_f1))\n",
    "print(\"Total models with optimal F1:\", len(hparams_max_f1))\n",
    "print()\n",
    "print(hparams_max_f1)\n",
    "\n",
    "max_accuracies = max(accuracies)\n",
    "max_accuracies_indices = [i for i,j in enumerate(accuracies) if j == max_accuracies]\n",
    "hparams_max_accuracies = [search_space[i] for i in max_accuracies_indices]\n",
    "print(\"\\n--- HYPERPARAMETERS FOR MAX ACCURACY OF {}: ---\\n\".format(max_accuracies))\n",
    "print(\"Total models with optimal accuracy:\", len(hparams_max_accuracies))\n",
    "print()\n",
    "print(hparams_max_accuracies)\n",
    "\n",
    "max_precision = max(precisions)\n",
    "max_precision_indices = [i for i,j in enumerate(precisions) if j == max_precision]\n",
    "hparams_max_precision = [search_space[i] for i in max_precision_indices]\n",
    "print(\"\\n--- HYPERPARAMETERS FOR MAX PRECISION OF {}: ---\\n\".format(max_precision))\n",
    "print(\"Total models with optimal precision:\", len(hparams_max_precision))\n",
    "print()\n",
    "print(hparams_max_precision)\n",
    "\n",
    "max_roc = max(rocs)\n",
    "max_roc_indices = [i for i,j in enumerate(rocs) if j == max_roc]\n",
    "hparams_max_roc = [search_space[i] for i in max_roc_indices]\n",
    "print(\"\\n--- HYPERPARAMETERS FOR MAX AUC ROC OF {}: ---\\n\".format(max_roc))\n",
    "print(\"Total models with optimal AUC ROC:\", len(hparams_max_roc))\n",
    "print()\n",
    "print(hparams_max_roc)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cc738a4b-d096-4988-aade-b696c09d3342",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total models at intersection: 5\n",
      "{(100, 0.4, 0.01), (250, 0.4, 0.01), (50, 0.4, 0.01), (100, 0.4, 0.1), (250, 0.4, 0.1)}\n"
     ]
    }
   ],
   "source": [
    "# Assess intersection of best models.\n",
    "hparams_max = [hparams_max_accuracies, \n",
    "               hparams_max_f1, \n",
    "               hparams_max_recall,\n",
    "               hparams_max_roc,\n",
    "               hparams_max_precision]\n",
    "best_intersect = set.intersection(*map(set, hparams_max))\n",
    "print(\"Total models at intersection:\", len(best_intersect))\n",
    "print(best_intersect)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d67502a0-4a96-4341-a64c-1198b178422a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimal hyperparameters = (50, 0.4, 0.01)\n"
     ]
    }
   ],
   "source": [
    "# Hand select given multiple optimal models.\n",
    "optimal_hparams = list(best_intersect)[2]\n",
    "print(\"Optimal hyperparameters =\", optimal_hparams)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "528994b1-b4b0-4943-9328-d420ae2ba5f0",
   "metadata": {},
   "source": [
    "## Train optimal labeling model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "43fd362f-9454-4b70-9bfe-06be99bc2ec7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████| 50/50 [00:00<00:00, 910.30epoch/s]\n",
      "WARNING:root:Metrics calculated over data points with non-abstain labels only\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Majority Model F1:        87.3%\n",
      "Majority Model Accuracy:  86.5%\n",
      "Majority Model Precision: 80.9%\n",
      "Majority Model Recall:    94.8%\n",
      "Majority Model AUC ROC:   86.6%\n",
      "Majority Model Coverage:  77.2%\n",
      "--------------------------------------\n",
      "Label Model F1:           82.5%\n",
      "Label Model Accuracy:     84.8%\n",
      "Label Model Precision:    73.0%\n",
      "Label Model Recall:       94.8%\n",
      "Label Model AUC ROC:      90.3%\n",
      "Label Model Coverage:     100.0%\n"
     ]
    }
   ],
   "source": [
    "# Extract hyperparameter values.\n",
    "n_epochs = optimal_hparams[0]\n",
    "l2 = optimal_hparams[1]\n",
    "lr = optimal_hparams[2]\n",
    "    \n",
    "# Label model.\n",
    "label_model = LabelModel(cardinality = 2, verbose = True)\n",
    "label_model.fit(L_train = L_train, \n",
    "                #Y_dev = y_val,\n",
    "                #class_balance = [0.7, 0.3], \n",
    "                n_epochs = n_epochs, \n",
    "                l2 = l2,\n",
    "                lr = lr,\n",
    "                optimizer = \"sgd\",\n",
    "                seed = seed)\n",
    "\n",
    "# Compute model performance metrics.\n",
    "label_scores = label_model.score(L = L_test,\n",
    "                                 Y = y_test, \n",
    "                                 tie_break_policy = \"abstain\",\n",
    "                                 metrics = [\"f1\", \"accuracy\", \"precision\", \n",
    "                                            \"recall\", \"roc_auc\", \"coverage\"])\n",
    "    \n",
    "    \n",
    "\n",
    "# Compare model performance metrics.\n",
    "majority_f1 = majority_scores.get(\"f1\")\n",
    "majority_acc = majority_scores.get(\"accuracy\")\n",
    "majority_prec = majority_scores.get(\"precision\")\n",
    "majority_rec = majority_scores.get(\"recall\")\n",
    "majority_roc = majority_scores.get(\"roc_auc\")\n",
    "majority_cov = majority_scores.get(\"coverage\")\n",
    "print(f\"{'Majority Model F1:':<25} {majority_f1 * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Accuracy:':<25} {majority_acc * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Precision:':<25} {majority_prec * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Recall:':<25} {majority_rec * 100:.1f}%\")\n",
    "print(f\"{'Majority Model AUC ROC:':<25} {majority_roc * 100:.1f}%\")\n",
    "print(f\"{'Majority Model Coverage:':<25} {majority_cov * 100:.1f}%\")\n",
    "print(\"--------------------------------------\")\n",
    "label_f1 = label_scores.get(\"f1\")\n",
    "label_acc = label_scores.get(\"accuracy\")\n",
    "label_prec = label_scores.get(\"precision\")\n",
    "label_rec = label_scores.get(\"recall\")\n",
    "label_roc = label_scores.get(\"roc_auc\")\n",
    "label_cov = label_scores.get(\"coverage\")\n",
    "print(f\"{'Label Model F1:':<25} {label_f1 * 100:.1f}%\")\n",
    "print(f\"{'Label Model Accuracy:':<25} {label_acc * 100:.1f}%\")\n",
    "print(f\"{'Label Model Precision:':<25} {label_prec * 100:.1f}%\")\n",
    "print(f\"{'Label Model Recall:':<25} {label_rec * 100:.1f}%\")\n",
    "print(f\"{'Label Model AUC ROC:':<25} {label_roc * 100:.1f}%\")\n",
    "print(f\"{'Label Model Coverage:':<25} {label_cov * 100:.1f}%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "458bbcdc-d358-42d7-83f9-54aa69227eae",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Export labeling function matrices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "cdcd35cd-e23a-4527-9367-117f6db63181",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRAIN:\n",
      " 0    844\n",
      "1    812\n",
      "dtype: int64\n",
      "VAL:\n",
      " 0    98\n",
      "1    86\n",
      "dtype: int64\n",
      "TEST:\n",
      " 0    234\n",
      "1    226\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Get predicted labels.\n",
    "pred_train = label_model.predict(L_train)\n",
    "pred_val = label_model.predict(L_val)\n",
    "pred_test = label_model.predict(L_test)\n",
    "\n",
    "print(\"TRAIN:\\n\", pd.Series(pred_train).value_counts())\n",
    "print(\"VAL:\\n\", pd.Series(pred_val).value_counts())\n",
    "print(\"TEST:\\n\", pd.Series(pred_test).value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "8c146ebd-cf60-45cf-b30e-e7bb684b1be3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-1    844\n",
      " 1    812\n",
      "dtype: int64\n",
      "-1    234\n",
      " 1    226\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Replace -1 with 0 (abstain = 0 in our model).\n",
    "# Replace 0 with -1 (-1 is class label in our model).\n",
    "pred_train = np.where(pred_train == -1, 5, pred_train)\n",
    "pred_train = np.where(pred_train == 0, -1, pred_train)\n",
    "pred_train = np.where(pred_train == 5, 0, pred_train)\n",
    "\n",
    "pred_val = np.where(pred_val == -1, 5, pred_val)\n",
    "pred_val = np.where(pred_val == 0, -1, pred_val)\n",
    "pred_val = np.where(pred_val == 5, 0, pred_val)\n",
    "\n",
    "pred_test = np.where(pred_test == -1, 5, pred_test)\n",
    "pred_test = np.where(pred_test == 0, -1, pred_test)\n",
    "pred_test = np.where(pred_test == 5, 0, pred_test)\n",
    "\n",
    "print(pd.Series(pred_train).value_counts())\n",
    "print(pd.Series(pred_test).value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "a509fe50-e909-4342-be29-644228aa795b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>Label</th>\n",
       "      <th>Snorkel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1651</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1652</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1653</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1654</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1655</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1656 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      0  1  2  Label  Snorkel\n",
       "0     0  0  1      1        1\n",
       "1     0  0  1     -1        1\n",
       "2     0 -1  1     -1       -1\n",
       "3     0  0  1      1        1\n",
       "4     0  0 -1     -1       -1\n",
       "...  .. .. ..    ...      ...\n",
       "1651  0  0  1      1        1\n",
       "1652  0 -1  1     -1       -1\n",
       "1653  0 -1 -1     -1       -1\n",
       "1654  0 -1 -1     -1       -1\n",
       "1655  0  0  1     -1        1\n",
       "\n",
       "[1656 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>Label</th>\n",
       "      <th>Snorkel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>179</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>181</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>183</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>184 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0  1  2  Label  Snorkel\n",
       "0    0  0 -1     -1       -1\n",
       "1    0 -1  1     -1       -1\n",
       "2    0  0  1      1        1\n",
       "3    0  0  1      1        1\n",
       "4    0 -1 -1     -1       -1\n",
       "..  .. .. ..    ...      ...\n",
       "179  0  0  1      1        1\n",
       "180  0  0 -1     -1       -1\n",
       "181  0 -1 -1     -1       -1\n",
       "182  0 -1  1     -1       -1\n",
       "183  0 -1 -1     -1       -1\n",
       "\n",
       "[184 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>Label</th>\n",
       "      <th>Snorkel</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>460 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     0  1  2  Label  Snorkel\n",
       "0    0  0  1      1        1\n",
       "1    0  0 -1     -1       -1\n",
       "2    0  0  1     -1        1\n",
       "3    0  0  1      1        1\n",
       "4    0  0  1      1        1\n",
       "..  .. .. ..    ...      ...\n",
       "455  0  0  1      1        1\n",
       "456  0  0 -1     -1       -1\n",
       "457  0 -1 -1     -1       -1\n",
       "458  0  0  1      1        1\n",
       "459  0  0  1     -1        1\n",
       "\n",
       "[460 rows x 5 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Replace -1 with 0 (abstain = 0 in our model).\n",
    "# Replace 0 with -1 (-1 is class label in our model).\n",
    "\n",
    "################################################################\n",
    "# TRAIN.\n",
    "L_train_new = np.where(L_train == -1, 5, L_train)\n",
    "L_train_new = np.where(L_train_new == 0, -1, L_train_new)\n",
    "L_train_new = np.where(L_train_new == 5, 0, L_train_new)\n",
    "y_train_new = np.where(y_train == 0, -1, y_train)\n",
    "\n",
    "df_L_train = pd.DataFrame(L_train_new)\n",
    "df_L_train[\"Label\"] = y_train_new\n",
    "df_L_train[\"Snorkel\"] = pred_train\n",
    "display(df_L_train)\n",
    "################################################################\n",
    "\n",
    "################################################################\n",
    "# VAL.\n",
    "L_val_new = np.where(L_val == -1, 5, L_val)\n",
    "L_val_new = np.where(L_val_new == 0, -1, L_val_new)\n",
    "L_val_new = np.where(L_val_new == 5, 0, L_val_new)\n",
    "y_val_new = np.where(y_val == 0, -1, y_val)\n",
    "\n",
    "df_L_val = pd.DataFrame(L_val_new)\n",
    "df_L_val[\"Label\"] = y_val_new\n",
    "df_L_val[\"Snorkel\"] = pred_val\n",
    "display(df_L_val)\n",
    "################################################################\n",
    "\n",
    "################################################################\n",
    "# TEST.\n",
    "L_test_new = np.where(L_test == -1, 5, L_test)\n",
    "L_test_new = np.where(L_test_new == 0, -1, L_test_new)\n",
    "L_test_new = np.where(L_test_new == 5, 0, L_test_new)\n",
    "y_test_new = np.where(y_test == 0, -1, y_test)\n",
    "\n",
    "df_L_test = pd.DataFrame(L_test_new)\n",
    "df_L_test[\"Label\"] = y_test_new\n",
    "df_L_test[\"Snorkel\"] = pred_test\n",
    "display(df_L_test)\n",
    "################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "03b4e958-237b-42e5-ab53-956ffd286912",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export labeling function matrices for external experiments.\n",
    "df_L_train.to_csv(\"rna_tuned_lf_matrix_train.csv\", index = False)\n",
    "df_L_val.to_csv(\"rna_tuned_lf_matrix_val.csv\", index = False)\n",
    "df_L_test.to_csv(\"rna_tuned_lf_matrix_test.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "263f70b2-5506-4b84-bea7-87e461d9148b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Export raw dataframes, split by train / val / test.\n",
    "df_train.to_csv(\"rna_tuned_train.csv\", index = False)\n",
    "df_val.to_csv(\"rna_tuned_val.csv\", index = False)\n",
    "df_test.to_csv(\"rna_tuned_test.csv\", index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62a338c-04ed-4a05-bce4-a4b771015895",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
